{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Coref_CRAFT.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "toc_visible": true
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "accelerator": "GPU"
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rKsIN3H3QPBl"
      },
      "source": [
        "### System Installation\n",
        "Installing the coref tool from mandarjoshi90 along with tensorflow.\n",
        "\n",
        "\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cVxSDHkTAhjr",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "ee77726e-b65a-4a0b-b08a-5be3aa176fb6"
      },
      "source": [
        "!nvidia-smi --query-gpu=gpu_name,driver_version,memory.total --format=csv"
      ],
      "execution_count": 1,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "name, driver_version, memory.total [MiB]\n",
            "Tesla T4, 460.32.03, 15109 MiB\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "g07Sx5jqQbPw",
        "outputId": "acb9f70d-fff4-49a8-b855-2176ecaa161f"
      },
      "source": [
        "! git clone https://github.com/mandarjoshi90/coref.git"
      ],
      "execution_count": 2,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Cloning into 'coref'...\n",
            "remote: Enumerating objects: 734, done.\u001b[K\n",
            "remote: Counting objects: 100% (6/6), done.\u001b[K\n",
            "remote: Compressing objects: 100% (6/6), done.\u001b[K\n",
            "remote: Total 734 (delta 2), reused 0 (delta 0), pack-reused 728\u001b[K\n",
            "Receiving objects: 100% (734/734), 4.17 MiB | 20.55 MiB/s, done.\n",
            "Resolving deltas: 100% (441/441), done.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "O4mHoDEKQdLu",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "d717891d-b78b-4602-ebb4-461725b40e59"
      },
      "source": [
        "%cd coref\n",
        "! sed 's/MarkupSafe==1.0/MarkupSafe==1.1.1/; s/scikit-learn==0.19.1/scikit-learn==0.21/; s/scipy==1.0.0/scipy==1.6.2/' < requirements.txt > tmp\n",
        "! mv tmp requirements.txt\n",
        "\n",
        "! sed 's/.D.GLIBCXX.USE.CXX11.ABI.0//' < setup_all.sh  > tmp\n",
        "! mv tmp setup_all.sh \n",
        "! chmod u+x setup_all.sh "
      ],
      "execution_count": 3,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/content/coref\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "cABghgMnQgkw",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "edc47340-1797-4078-d90e-9a61e5d85f61"
      },
      "source": [
        "%tensorflow_version 2.x\n",
        "! pip uninstall -y tensorflow\n",
        "! pip install -r requirements.txt --log install-log.txt -q\n",
        "! ./setup_all.sh"
      ],
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Uninstalling tensorflow-2.4.1:\n",
            "  Successfully uninstalled tensorflow-2.4.1\n",
            "\u001b[K     |████████████████████████████████| 102kB 11.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 1.2MB 40.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 163kB 53.5MB/s \n",
            "\u001b[K     |████████████████████████████████| 6.6MB 27.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 552kB 53.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 61kB 9.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 2.2MB 35.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 266kB 55.5MB/s \n",
            "\u001b[K     |████████████████████████████████| 890kB 38.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 133kB 57.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 153kB 59.8MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 7.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 7.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 92kB 13.6MB/s \n",
            "\u001b[K     |████████████████████████████████| 20.3MB 1.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 2.1MB 37.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 1.2MB 42.1MB/s \n",
            "\u001b[K     |████████████████████████████████| 430kB 55.6MB/s \n",
            "\u001b[K     |████████████████████████████████| 71kB 10.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 256kB 59.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 71kB 11.2MB/s \n",
            "\u001b[K     |████████████████████████████████| 61kB 8.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 194kB 55.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 512kB 48.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 61kB 9.8MB/s \n",
            "\u001b[K     |████████████████████████████████| 6.7MB 48.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 27.4MB 118kB/s \n",
            "\u001b[K     |████████████████████████████████| 3.2MB 42.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 491kB 52.8MB/s \n",
            "\u001b[K     |████████████████████████████████| 377.1MB 48kB/s \n",
            "\u001b[K     |████████████████████████████████| 748.9MB 19kB/s \n",
            "\u001b[K     |████████████████████████████████| 8.8MB 28.7MB/s \n",
            "\u001b[K     |████████████████████████████████| 327kB 58.9MB/s \n",
            "\u001b[K     |████████████████████████████████| 51kB 8.4MB/s \n",
            "\u001b[K     |████████████████████████████████| 256kB 57.3MB/s \n",
            "\u001b[K     |████████████████████████████████| 4.1MB 48.7MB/s \n",
            "\u001b[?25h  Building wheel for absl-py (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for gast (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for h5py (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for html5lib (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for JPype1 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for mmh3 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for msgpack-python (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for psycopg2 (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for pycparser (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for pyhocon (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for wrapt (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Building wheel for PyYAML (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "\u001b[31mERROR: fancyimpute 0.4.3 requires tensorflow, which is not installed.\u001b[0m\n",
            "\u001b[31mERROR: torchtext 0.9.1 has requirement torch==1.8.1, but you'll have torch 1.2.0 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: tensorflow-probability 0.12.1 has requirement gast>=0.3.2, but you'll have gast 0.2.2 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: tensorflow-metadata 0.30.0 has requirement absl-py<0.13,>=0.9, but you'll have absl-py 0.7.1 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: pyasn1-modules 0.2.8 has requirement pyasn1<0.5.0,>=0.4.6, but you'll have pyasn1 0.4.2 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: pandas 1.1.5 has requirement python-dateutil>=2.7.3, but you'll have python-dateutil 2.6.1 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: googleapis-common-protos 1.53.0 has requirement protobuf>=3.12.0, but you'll have protobuf 3.9.1 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: google-colab 1.0.0 has requirement astor~=0.8.1, but you'll have astor 0.8.0 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: google-colab 1.0.0 has requirement six~=1.15.0, but you'll have six 1.12.0 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: google-api-python-client 1.12.8 has requirement six<2dev,>=1.13.0, but you'll have six 1.12.0 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: google-api-core 1.26.3 has requirement protobuf>=3.12.0, but you'll have protobuf 3.9.1 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: google-api-core 1.26.3 has requirement six>=1.13.0, but you'll have six 1.12.0 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: flask 1.1.2 has requirement Jinja2>=2.10.1, but you'll have jinja2 2.10 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: flask 1.1.2 has requirement Werkzeug>=0.15, but you'll have werkzeug 0.14.1 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: fbprophet 0.7.1 has requirement python-dateutil>=2.8.0, but you'll have python-dateutil 2.6.1 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: datascience 0.10.6 has requirement folium==0.2.1, but you'll have folium 0.8.3 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: bokeh 2.3.2 has requirement pillow>=7.1.0, but you'll have pillow 6.1.0 which is incompatible.\u001b[0m\n",
            "\u001b[31mERROR: albumentations 0.1.12 has requirement imgaug<0.2.7,>=0.2.5, but you'll have imgaug 0.2.9 which is incompatible.\u001b[0m\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
            "  from ._conv import register_converters as _register_converters\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
            "  from ._conv import register_converters as _register_converters\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "bSfzbpk2oVsQ",
        "outputId": "49430d2c-e8a2-467b-afea-5e06f2346fc6"
      },
      "source": [
        "import nltk\n",
        "nltk.download('punkt')"
      ],
      "execution_count": 5,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "[nltk_data] Downloading package punkt to /root/nltk_data...\n",
            "[nltk_data]   Unzipping tokenizers/punkt.zip.\n"
          ],
          "name": "stdout"
        },
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "True"
            ]
          },
          "metadata": {
            "tags": []
          },
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "OtaUoOBJkNDw"
      },
      "source": [
        "### Specifying Input\n",
        "\n",
        "Input and Model"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "tuq5EkiiqkpW",
        "outputId": "ff8aef8d-9534-4460-e048-c232230ef255"
      },
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')"
      ],
      "execution_count": 6,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Mounted at /content/gdrive\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "EaFGGbm2QN7r"
      },
      "source": [
        "genre = \"nw\"\n",
        "# The Ontonotes data for training the model contains text from several sources\n",
        "# of very different styles. You need to specify the most suitable one out of:\n",
        "# \"bc\": broadcast conversation\n",
        "# \"bn\": broadcast news\n",
        "# \"mz\": magazine\n",
        "# \"nw\": newswire\n",
        "# \"pt\": Bible text\n",
        "# \"tc\": telephone conversation\n",
        "# \"wb\": web data\n",
        "\n",
        "model_name = \"spanbert_base\"\n",
        "# The fine-tuned model to use. Options are:\n",
        "# bert_base\n",
        "# spanbert_base\n",
        "# bert_large\n",
        "# spanbert_large"
      ],
      "execution_count": 7,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "BQ163Z2uQevj"
      },
      "source": [
        "import os\n",
        "os.environ['data_dir'] = \"./data\"\n",
        "os.environ['CHOSEN_MODEL'] = model_name"
      ],
      "execution_count": 8,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "xi0-5VN-kr6N"
      },
      "source": [
        "Downloading the selected model."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "klQ48SSkQiWG",
        "outputId": "f42254a2-f1d9-482e-aebc-a889f37cebd2"
      },
      "source": [
        "! ./download_pretrained.sh $CHOSEN_MODEL"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Downloading spanbert_base\n",
            "--2021-05-16 17:24:36--  http://nlp.cs.washington.edu/pair2vec/spanbert_base.tar.gz\n",
            "Resolving nlp.cs.washington.edu (nlp.cs.washington.edu)... 128.208.3.120, 2607:4000:200:12::78\n",
            "Connecting to nlp.cs.washington.edu (nlp.cs.washington.edu)|128.208.3.120|:80... connected.\n",
            "HTTP request sent, awaiting response... 200 OK\n",
            "Length: 1633726311 (1.5G) [application/x-gzip]\n",
            "Saving to: ‘./data/spanbert_base.tar.gz’\n",
            "\n",
            "spanbert_base.tar.g  57%[==========>         ] 896.47M  18.6MB/s    eta 38s    "
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "MDl_lF2EkyCx"
      },
      "source": [
        "Process the data to be in the required input format:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "O-rCm36PQlaB"
      },
      "source": [
        "from bert import tokenization\n",
        "import json\n",
        "\n",
        "\n",
        "def tokenize(data):\n",
        "  data = {\n",
        "      'doc_key': genre,\n",
        "      'sentences': [[\"[CLS]\"]],\n",
        "      'speakers': [[\"[SPL]\"]],\n",
        "      'clusters': [],\n",
        "      'sentence_map': [0],\n",
        "      'subtoken_map': [0],\n",
        "  }\n",
        "\n",
        "  # Determine Max Segment\n",
        "  max_segment = None\n",
        "  for line in open('experiments.conf'):\n",
        "      if line.startswith(model_name):\n",
        "          max_segment = True\n",
        "      elif line.strip().startswith(\"max_segment_len\"):\n",
        "          if max_segment:\n",
        "              max_segment = int(line.strip().split()[-1])\n",
        "              break\n",
        "\n",
        "  tokenizer = tokenization.FullTokenizer(vocab_file=\"cased_config_vocab/vocab.txt\", do_lower_case=False)\n",
        "  subtoken_num = 0\n",
        "  for sent_num, line in enumerate(text):\n",
        "      raw_tokens = line.split()\n",
        "      tokens = tokenizer.tokenize(line)\n",
        "      if len(tokens) + len(data['sentences'][-1]) >= max_segment:\n",
        "          data['sentences'][-1].append(\"[SEP]\")\n",
        "          data['sentences'].append([\"[CLS]\"])\n",
        "          data['speakers'][-1].append(\"[SPL]\")\n",
        "          data['speakers'].append([\"[SPL]\"])\n",
        "          data['sentence_map'].append(sent_num - 1)\n",
        "          data['subtoken_map'].append(subtoken_num - 1)\n",
        "          data['sentence_map'].append(sent_num)\n",
        "          data['subtoken_map'].append(subtoken_num)\n",
        "\n",
        "      ctoken = raw_tokens[0]\n",
        "      cpos = 0\n",
        "      for token in tokens:\n",
        "          data['sentences'][-1].append(token)\n",
        "          data['speakers'][-1].append(\"-\")\n",
        "          data['sentence_map'].append(sent_num)\n",
        "          data['subtoken_map'].append(subtoken_num)\n",
        "          \n",
        "          if token.startswith(\"##\"):\n",
        "              token = token[2:]\n",
        "          if len(ctoken) == len(token):\n",
        "              subtoken_num += 1\n",
        "              cpos += 1\n",
        "              if cpos < len(raw_tokens):\n",
        "                  ctoken = raw_tokens[cpos]\n",
        "          else:\n",
        "              ctoken = ctoken[len(token):]\n",
        "\n",
        "  data['sentences'][-1].append(\"[SEP]\")\n",
        "  data['speakers'][-1].append(\"[SPL]\")\n",
        "  data['sentence_map'].append(sent_num - 1)\n",
        "  data['subtoken_map'].append(subtoken_num - 1)\n",
        "\n",
        "  return data\n",
        "\n",
        "def chunks(lst, n):\n",
        "    \"\"\"Yield successive n-sized chunks from lst.\"\"\"\n",
        "    for i in range(0, len(lst), n):\n",
        "        yield lst[i:i + n]"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "cuC9mlsWkTuP"
      },
      "source": [
        "## Prediction & Evaluation\n"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "JDJeLUoSV7kK"
      },
      "source": [
        "%mkdir ./data/in\n",
        "%mkdir ./data/out"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "uFCJLo2zPDPP"
      },
      "source": [
        "import subprocess\n",
        "\n",
        "tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')\n",
        "filename = \"/content/gdrive/MyDrive/CRAFT-txt/dev/17194222.txt\"\n",
        "\n",
        "\n",
        "\n",
        "text = []\n",
        "data = []\n",
        "\n",
        "with open(filename) as f_in:\n",
        "  \n",
        "  paragraphs = [line for line in f_in.read().split('\\n') if line]\n",
        "  \n",
        "  for paragraph in paragraphs:\n",
        "    sentences = tokenizer.tokenize(paragraph)\n",
        "    text.extend(sentences)\n",
        "                \n",
        "chs = list(chunks(text, 100))\n",
        "\n",
        "N = len(chs)\n",
        "\n",
        "for i in range(0, N):\n",
        "  \n",
        "  text = chs[i]\n",
        "  data.append(tokenize(text))\n",
        "\n",
        "  file1 = \"./data/in/\" + str(i) + \".json\"\n",
        "  file2 = \"./data/out/_\" + str(i) + \".json\"\n",
        "\n",
        "  with open(file1, 'w') as out:\n",
        "    json.dump(data[i], out, sort_keys=True)\n",
        "\n",
        "  subprocess.call(['python', 'predict.py', 'spanbert_base', file1, file2])\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ZrQrj-PDW0h4"
      },
      "source": [
        "import subprocess\n",
        "tokenizer = nltk.data.load('tokenizers/punkt/english.pickle')\n",
        "\n",
        "for file in [f for f in os.listdir('/content/gdrive/MyDrive/CRAFT-txt/dev') if f.endswith('.txt')]:\n",
        "\n",
        "  text = []\n",
        "  data = []\n",
        "\n",
        "  with open(filename) as f_in:\n",
        "    \n",
        "    paragraphs = [line for line in f_in.read().split('\\n') if line]\n",
        "  \n",
        "    for paragraph in paragraphs:\n",
        "      sentences = tokenizer.tokenize(paragraph)\n",
        "      text.extend(sentences)\n",
        "                \n",
        "  chs = list(chunks(text, 100))\n",
        "\n",
        "  N = len(chs)\n",
        "\n",
        "  for i in range(0, N):\n",
        "    \n",
        "    text = chs[i]\n",
        "    data.append(tokenize(text))\n",
        "\n",
        "    file1 = \"./data/in/\"  + file[0:-4] + \"_\" + str(i) + \".json\"\n",
        "    file2 = \"./data/out/\" + file[0:-4] + \"_\" + str(i) + \".jsonlines\"\n",
        "\n",
        "    with open(file1, 'w') as out:\n",
        "      json.dump(data[i], out, sort_keys=True)\n",
        "    \n",
        "    subprocess.call(['python', 'predict.py', 'spanbert_base', file1, file2])\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "zniQuqNGQnOk",
        "outputId": "9ab328fc-b3bb-4c1c-e253-1b9de439d3c1"
      },
      "source": [
        "! gpu=0 python predict.py $CHOSEN_MODEL ./data/in_0.json ./data/out_0.json"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:516: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:517: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:518: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:519: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:520: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.\n",
            "  from ._conv import register_converters as _register_converters\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:541: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint8 = np.dtype([(\"qint8\", np.int8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:542: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint8 = np.dtype([(\"quint8\", np.uint8, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:543: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint16 = np.dtype([(\"qint16\", np.int16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:544: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_quint16 = np.dtype([(\"quint16\", np.uint16, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:545: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  _np_qint32 = np.dtype([(\"qint32\", np.int32, 1)])\n",
            "/usr/local/lib/python3.7/dist-packages/tensorboard/compat/tensorflow_stub/dtypes.py:550: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.\n",
            "  np_resource = np.dtype([(\"resource\", np.ubyte, 1)])\n",
            "WARNING: Logging before flag parsing goes to stderr.\n",
            "W0514 14:30:05.058650 139705853998976 deprecation_wrapper.py:119] From /content/coref/coref_ops.py:11: The name tf.NotDifferentiable is deprecated. Please use tf.no_gradient instead.\n",
            "\n",
            "/usr/local/lib/python3.7/dist-packages/sklearn/utils/linear_assignment_.py:21: DeprecationWarning: The linear_assignment_ module is deprecated in 0.21 and will be removed from 0.23. Use scipy.optimize.linear_sum_assignment instead.\n",
            "  DeprecationWarning)\n",
            "W0514 14:30:05.100141 139705853998976 deprecation_wrapper.py:119] From /content/coref/bert/optimization.py:87: The name tf.train.Optimizer is deprecated. Please use tf.compat.v1.train.Optimizer instead.\n",
            "\n",
            "W0514 14:30:05.994991 139705853998976 lazy_loader.py:50] \n",
            "The TensorFlow contrib module will not be included in TensorFlow 2.0.\n",
            "For more information, please see:\n",
            "  * https://github.com/tensorflow/community/blob/master/rfcs/20180907-contrib-sunset.md\n",
            "  * https://github.com/tensorflow/addons\n",
            "  * https://github.com/tensorflow/io (for I/O related ops)\n",
            "If you depend on functionality not listed there, please file an issue.\n",
            "\n",
            "Running experiment: spanbert_base\n",
            "data_dir = \"/sdb/data/new_coref\"\n",
            "model_type = \"independent\"\n",
            "max_top_antecedents = 50\n",
            "max_training_sentences = 3\n",
            "top_span_ratio = 0.4\n",
            "max_num_speakers = 20\n",
            "max_segment_len = 384\n",
            "bert_learning_rate = 2e-05\n",
            "task_learning_rate = 0.0001\n",
            "num_docs = 2802\n",
            "dropout_rate = 0.3\n",
            "ffnn_size = 3000\n",
            "ffnn_depth = 1\n",
            "num_epochs = 20\n",
            "feature_size = 20\n",
            "max_span_width = 30\n",
            "use_metadata = true\n",
            "use_features = true\n",
            "use_segment_distance = true\n",
            "model_heads = true\n",
            "coref_depth = 2\n",
            "coarse_to_fine = true\n",
            "fine_grained = true\n",
            "use_prior = true\n",
            "train_path = \"./data/train.english.384.jsonlines\"\n",
            "eval_path = \"./data/dev.english.384.jsonlines\"\n",
            "conll_eval_path = \"./data/dev.english.v4_gold_conll\"\n",
            "single_example = true\n",
            "genres = [\n",
            "  \"bc\"\n",
            "  \"bn\"\n",
            "  \"mz\"\n",
            "  \"nw\"\n",
            "  \"pt\"\n",
            "  \"tc\"\n",
            "  \"wb\"\n",
            "]\n",
            "eval_frequency = 1000\n",
            "report_frequency = 100\n",
            "log_root = \"./data\"\n",
            "adam_eps = 1e-06\n",
            "task_optimizer = \"adam\"\n",
            "bert_config_file = \"./data/spanbert_base/bert_config.json\"\n",
            "vocab_file = \"./data/spanbert_base/vocab.txt\"\n",
            "tf_checkpoint = \"./data/spanbert_base/model.max.ckpt\"\n",
            "init_checkpoint = \"./data/spanbert_base/model.max.ckpt\"\n",
            "log_dir = \"./data/spanbert_base\"\n",
            "W0514 14:30:06.056577 139705853998976 deprecation_wrapper.py:119] From /content/coref/bert/modeling.py:92: The name tf.gfile.GFile is deprecated. Please use tf.io.gfile.GFile instead.\n",
            "\n",
            "W0514 14:30:06.130739 139705853998976 deprecation_wrapper.py:119] From /content/coref/independent.py:48: The name tf.placeholder is deprecated. Please use tf.compat.v1.placeholder instead.\n",
            "\n",
            "W0514 14:30:06.136145 139705853998976 deprecation_wrapper.py:119] From /content/coref/independent.py:50: The name tf.PaddingFIFOQueue is deprecated. Please use tf.queue.PaddingFIFOQueue instead.\n",
            "\n",
            "W0514 14:30:06.139153 139705853998976 deprecation.py:323] From /content/coref/bert/modeling.py:158: to_float (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use `tf.cast` instead.\n",
            "W0514 14:30:06.150894 139705853998976 deprecation_wrapper.py:119] From /content/coref/bert/modeling.py:175: The name tf.variable_scope is deprecated. Please use tf.compat.v1.variable_scope instead.\n",
            "\n",
            "W0514 14:30:06.151029 139705853998976 deprecation_wrapper.py:119] From /content/coref/bert/modeling.py:175: The name tf.AUTO_REUSE is deprecated. Please use tf.compat.v1.AUTO_REUSE instead.\n",
            "\n",
            "W0514 14:30:06.212574 139705853998976 deprecation.py:506] From /content/coref/bert/modeling.py:362: calling dropout (from tensorflow.python.ops.nn_ops) with keep_prob is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Please use `rate` instead of `keep_prob`. Rate should be set to `rate = 1 - keep_prob`.\n",
            "W0514 14:30:06.252979 139705853998976 deprecation.py:323] From /content/coref/bert/modeling.py:676: dense (from tensorflow.python.layers.core) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use keras.layers.dense instead.\n",
            "W0514 14:30:08.742758 139705853998976 deprecation.py:323] From /usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/array_ops.py:1354: add_dispatch_support.<locals>.wrapper (from tensorflow.python.ops.array_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use tf.where in 2.0, which has the same broadcast rule as np.where\n",
            "W0514 14:30:08.820300 139705853998976 deprecation.py:323] From /content/coref/independent.py:221: to_int32 (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use `tf.cast` instead.\n",
            "W0514 14:30:08.865341 139705853998976 deprecation_wrapper.py:119] From /content/coref/util.py:117: The name tf.nn.xw_plus_b is deprecated. Please use tf.compat.v1.nn.xw_plus_b instead.\n",
            "\n",
            "W0514 14:30:09.371789 139705853998976 deprecation_wrapper.py:119] From /content/coref/independent.py:60: The name tf.train.init_from_checkpoint is deprecated. Please use tf.compat.v1.train.init_from_checkpoint instead.\n",
            "\n",
            "**** Trainable Variables ****\n",
            "  name = bert/embeddings/word_embeddings:0, shape = (28996, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/embeddings/token_type_embeddings:0, shape = (2, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/embeddings/position_embeddings:0, shape = (512, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/embeddings/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/embeddings/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_0/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_1/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_2/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_3/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_4/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_5/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_6/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_7/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_8/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_9/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_10/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/self/query/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/self/query/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/self/key/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/self/key/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/self/value/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/self/value/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/output/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/attention/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/intermediate/dense/kernel:0, shape = (768, 3072), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/intermediate/dense/bias:0, shape = (3072,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/output/dense/kernel:0, shape = (3072, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/output/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/output/LayerNorm/beta:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/encoder/layer_11/output/LayerNorm/gamma:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = bert/pooler/dense/kernel:0, shape = (768, 768), *INIT_FROM_CKPT*\n",
            "  name = bert/pooler/dense/bias:0, shape = (768,), *INIT_FROM_CKPT*\n",
            "  name = span_width_embeddings:0, shape = (30, 20), *INIT_FROM_CKPT*\n",
            "  name = mention_word_attn/output_weights:0, shape = (768, 1), *INIT_FROM_CKPT*\n",
            "  name = mention_word_attn/output_bias:0, shape = (1,), *INIT_FROM_CKPT*\n",
            "  name = mention_scores/hidden_weights_0:0, shape = (2324, 3000), *INIT_FROM_CKPT*\n",
            "  name = mention_scores/hidden_bias_0:0, shape = (3000,), *INIT_FROM_CKPT*\n",
            "  name = mention_scores/output_weights:0, shape = (3000, 1), *INIT_FROM_CKPT*\n",
            "  name = mention_scores/output_bias:0, shape = (1,), *INIT_FROM_CKPT*\n",
            "  name = span_width_prior_embeddings:0, shape = (30, 20), *INIT_FROM_CKPT*\n",
            "  name = width_scores/hidden_weights_0:0, shape = (20, 3000), *INIT_FROM_CKPT*\n",
            "  name = width_scores/hidden_bias_0:0, shape = (3000,), *INIT_FROM_CKPT*\n",
            "  name = width_scores/output_weights:0, shape = (3000, 1), *INIT_FROM_CKPT*\n",
            "  name = width_scores/output_bias:0, shape = (1,), *INIT_FROM_CKPT*\n",
            "  name = genre_embeddings:0, shape = (7, 20), *INIT_FROM_CKPT*\n",
            "  name = src_projection/output_weights:0, shape = (2324, 2324), *INIT_FROM_CKPT*\n",
            "  name = src_projection/output_bias:0, shape = (2324,), *INIT_FROM_CKPT*\n",
            "  name = antecedent_distance_emb:0, shape = (10, 20), *INIT_FROM_CKPT*\n",
            "  name = output_weights:0, shape = (20, 1), *INIT_FROM_CKPT*\n",
            "  name = output_bias:0, shape = (1,), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/same_speaker_emb:0, shape = (2, 20), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/antecedent_distance_emb:0, shape = (10, 20), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/segment_distance/segment_distance_embeddings:0, shape = (3, 20), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/slow_antecedent_scores/hidden_weights_0:0, shape = (7052, 3000), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/slow_antecedent_scores/hidden_bias_0:0, shape = (3000,), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/slow_antecedent_scores/output_weights:0, shape = (3000, 1), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/slow_antecedent_scores/output_bias:0, shape = (1,), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/f/output_weights:0, shape = (4648, 2324), *INIT_FROM_CKPT*\n",
            "  name = coref_layer/f/output_bias:0, shape = (2324,), *INIT_FROM_CKPT*\n",
            "W0514 14:30:10.151337 139705853998976 deprecation_wrapper.py:119] From /content/coref/independent.py:74: The name tf.train.get_or_create_global_step is deprecated. Please use tf.compat.v1.train.get_or_create_global_step instead.\n",
            "\n",
            "W0514 14:30:10.156309 139705853998976 deprecation_wrapper.py:119] From /content/coref/optimization.py:13: The name tf.train.polynomial_decay is deprecated. Please use tf.compat.v1.train.polynomial_decay instead.\n",
            "\n",
            "W0514 14:30:10.160013 139705853998976 deprecation.py:323] From /usr/local/lib/python3.7/dist-packages/tensorflow/python/keras/optimizer_v2/learning_rate_schedule.py:409: div (from tensorflow.python.ops.math_ops) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Deprecated in favor of operator or tf.math.divide.\n",
            "W0514 14:30:10.177911 139705853998976 deprecation_wrapper.py:119] From /content/coref/optimization.py:64: The name tf.train.AdamOptimizer is deprecated. Please use tf.compat.v1.train.AdamOptimizer instead.\n",
            "\n",
            "bert:task 199 27\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/gradients_util.py:93: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.\n",
            "  \"Converting sparse IndexedSlices to a dense Tensor of unknown shape. \"\n",
            "/usr/local/lib/python3.7/dist-packages/tensorflow/python/ops/gradients_util.py:93: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.\n",
            "  \"Converting sparse IndexedSlices to a dense Tensor of unknown shape. \"\n",
            "2021-05-14 14:30:19.227100: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcuda.so.1\n",
            "2021-05-14 14:30:19.264466: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.265098: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: \n",
            "name: Tesla T4 major: 7 minor: 5 memoryClockRate(GHz): 1.59\n",
            "pciBusID: 0000:00:04.0\n",
            "2021-05-14 14:30:19.265169: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0\n",
            "2021-05-14 14:30:19.266619: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0\n",
            "2021-05-14 14:30:19.268017: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0\n",
            "2021-05-14 14:30:19.268382: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0\n",
            "2021-05-14 14:30:19.270000: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0\n",
            "2021-05-14 14:30:19.271407: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0\n",
            "2021-05-14 14:30:19.275281: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7\n",
            "2021-05-14 14:30:19.275541: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.276155: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.276686: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0\n",
            "2021-05-14 14:30:19.277087: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA\n",
            "2021-05-14 14:30:19.281846: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2199995000 Hz\n",
            "2021-05-14 14:30:19.282040: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x56244eb6b640 executing computations on platform Host. Devices:\n",
            "2021-05-14 14:30:19.282066: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>\n",
            "2021-05-14 14:30:19.496864: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.497961: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x5624552d8a80 executing computations on platform CUDA. Devices:\n",
            "2021-05-14 14:30:19.497998: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): Tesla T4, Compute Capability 7.5\n",
            "2021-05-14 14:30:19.498232: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.499008: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1640] Found device 0 with properties: \n",
            "name: Tesla T4 major: 7 minor: 5 memoryClockRate(GHz): 1.59\n",
            "pciBusID: 0000:00:04.0\n",
            "2021-05-14 14:30:19.499094: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0\n",
            "2021-05-14 14:30:19.499142: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0\n",
            "2021-05-14 14:30:19.499166: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcufft.so.10.0\n",
            "2021-05-14 14:30:19.499185: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcurand.so.10.0\n",
            "2021-05-14 14:30:19.499203: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusolver.so.10.0\n",
            "2021-05-14 14:30:19.499222: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcusparse.so.10.0\n",
            "2021-05-14 14:30:19.499242: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudnn.so.7\n",
            "2021-05-14 14:30:19.499368: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.500192: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:19.500834: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1763] Adding visible gpu devices: 0\n",
            "2021-05-14 14:30:19.500907: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcudart.so.10.0\n",
            "2021-05-14 14:30:22.788511: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1181] Device interconnect StreamExecutor with strength 1 edge matrix:\n",
            "2021-05-14 14:30:22.788566: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1187]      0 \n",
            "2021-05-14 14:30:22.788579: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1200] 0:   N \n",
            "2021-05-14 14:30:22.788805: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:22.789491: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1005] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero\n",
            "2021-05-14 14:30:22.790010: W tensorflow/core/common_runtime/gpu/gpu_bfc_allocator.cc:40] Overriding allow_growth setting because the TF_FORCE_GPU_ALLOW_GROWTH environment variable is set. Original config value was 0.\n",
            "2021-05-14 14:30:22.790055: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1326] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 13533 MB memory) -> physical GPU (device: 0, name: Tesla T4, pci bus id: 0000:00:04.0, compute capability: 7.5)\n",
            "Restoring from ./data/spanbert_base/model.max.ckpt\n",
            "W0514 14:30:30.464008 139705853998976 deprecation.py:323] From /usr/local/lib/python3.7/dist-packages/tensorflow/python/training/saver.py:1276: checkpoint_exists (from tensorflow.python.training.checkpoint_management) is deprecated and will be removed in a future version.\n",
            "Instructions for updating:\n",
            "Use standard file APIs to check for files with this prefix.\n",
            "2021-05-14 14:30:33.112483: I tensorflow/stream_executor/platform/default/dso_loader.cc:42] Successfully opened dynamic library libcublas.so.10.0\n",
            "Decoded 1 examples.\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "M9RcLDGBNWM7"
      },
      "source": [
        "### Using Evaulate\n",
        "currently not working"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "HmpPpAAX6zrK"
      },
      "source": [
        "! gpu=0 python evaluate.py $CHOSEN_MODEL"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "TI6OkNG-kXIQ"
      },
      "source": [
        "## Output"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rXpA3sW5QpSA",
        "outputId": "b3220173-4bf5-4ade-a917-823749af9cb9"
      },
      "source": [
        "output = json.load(open(\"./data/dev_2.jsonlines\"))\n",
        "\n",
        "comb_text = [word for sentence in output['sentences'] for word in sentence]\n",
        "\n",
        "def convert_mention(mention):\n",
        "    start = output['subtoken_map'][mention[0]]\n",
        "    end = output['subtoken_map'][mention[1]] + 1\n",
        "    nmention = (start, end)\n",
        "    mtext = ''.join(' '.join(comb_text[mention[0]:mention[1]+1]).split(\" ##\"))\n",
        "    return (nmention, mtext)\n",
        "\n",
        "seen = set()\n",
        "print('Clusters:')\n",
        "for cluster in output['predicted_clusters']:\n",
        "    mapped = []\n",
        "    for mention in cluster:\n",
        "        seen.add(tuple(mention))\n",
        "        mapped.append(convert_mention(mention))\n",
        "    print(mapped, end=\",\\n\")\n",
        "\n",
        "print('\\nMentions:')\n",
        "for mention in output['top_spans']:\n",
        "    if tuple(mention) in seen:\n",
        "        continue\n",
        "    print(convert_mention(mention), end=\",\\n\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "Clusters:\n",
            "[((0, 1), 'We'), ((24, 25), 'we'), ((90, 91), 'we'), ((251, 252), 'we'), ((397, 398), 'we'), ((622, 623), 'we'), ((698, 699), 'we'), ((723, 724), 'We'), ((930, 931), 'we'), ((993, 994), 'we'), ((1130, 1131), 'us'), ((1158, 1159), 'we'), ((1205, 1206), 'our'), ((1270, 1271), 'We'), ((1685, 1686), 'we'), ((1824, 1825), 'our'), ((1833, 1834), 'We'), ((1847, 1848), 'we'), ((1859, 1860), 'Our'), ((1920, 1921), 'we'), ((1947, 1948), 'we')],\n",
            "[((54, 57), 'The mean CV'), ((67, 71), 'this relatively low CV')],\n",
            "[((97, 114), 'the DBA / 1 and FVB / N strains at all four phases of CIA , including NC , PI , OA and CA'), ((126, 129), 'the two strains'), ((135, 137), 'both strains'), ((164, 167), 'these two strains'), ((207, 210), 'the two strains'), ((600, 604), 'these two gene clusters'), ((644, 646), 'both lists'), ((818, 821), 'the two strains'), ((1282, 1288), 'both genetically susceptible and resistant strains'), ((1295, 1297), 'both strains'), ((1446, 1449), 'the two strains'), ((1510, 1512), 'both strains'), ((1816, 1818), 'two strains'), ((1916, 1919), 'the two strains')],\n",
            "[((107, 108), 'CIA'), ((143, 145), 'the disease'), ((154, 155), 'CIA'), ((230, 231), 'CIA'), ((250, 251), 'CIA'), ((262, 263), 'CIA'), ((307, 309), 'the disease'), ((320, 322), 'the disease'), ((343, 344), 'CIA'), ((381, 382), 'CIA'), ((396, 397), 'CIA'), ((425, 426), 'CIA'), ((439, 440), 'CIA'), ((453, 455), 'the disease'), ((487, 488), 'CIA'), ((611, 612), 'CIA'), ((619, 620), 'CIA'), ((783, 784), 'CIA'), ((809, 810), 'CIA'), ((834, 835), 'CIA'), ((861, 862), 'CIA'), ((991, 992), 'CIA'), ((1178, 1179), 'CIA'), ((1219, 1220), 'CIA'), ((1229, 1230), 'CIA'), ((1246, 1247), 'CIA'), ((1269, 1270), 'CIA'), ((1293, 1294), 'CIA'), ((1390, 1391), 'CIA'), ((1439, 1440), 'CIA'), ((1475, 1476), 'CIA'), ((1508, 1509), 'CIA'), ((1534, 1535), 'CIA'), ((1620, 1621), 'CIA'), ((1857, 1858), 'CIA'), ((1933, 1934), 'CIA'), ((1965, 1966), 'CIA')],\n",
            "[((98, 99), 'DBA / 1'), ((151, 152), 'DBA / 1'), ((853, 855), 'the DBA / 1'), ((873, 876), 'the DBA / 1 strain'), ((1087, 1088), 'DBA / 1'), ((1381, 1382), 'DBA / 1'), ((1422, 1425), 'the DBA / 1 strain'), ((1465, 1468), 'the DBA / 1 strain'), ((1490, 1493), 'the DBA / 1 strain')],\n",
            "[((103, 114), 'all four phases of CIA , including NC , PI , OA and CA'), ((183, 186), 'the four phases'), ((211, 213), 'all phases')],\n",
            "[((200, 202), 'Twenty - one genes'), ((214, 217), 'these 21 genes')],\n",
            "[((110, 111), 'PI'), ((232, 233), 'PI'), ((270, 271), 'PI'), ((357, 358), 'PI'), ((360, 361), 'PI'), ((882, 883), 'PI'), ((1305, 1306), 'PI'), ((1335, 1336), 'PI'), ((1392, 1393), 'PI')],\n",
            "[((113, 114), 'CA'), ((235, 236), 'CA'), ((273, 274), 'CA'), ((362, 363), 'CA'), ((366, 367), 'CA'), ((1307, 1308), 'CA'), ((1337, 1338), 'CA'), ((1402, 1403), 'CA')],\n",
            "[((109, 110), 'NC'), ((278, 279), 'NC'), ((823, 824), 'NC'), ((881, 882), 'NC'), ((1391, 1392), 'NC')],\n",
            "[((285, 287), '102 genes'), ((293, 294), 'them')],\n",
            "[((332, 335), 'Only one gene'), ((345, 347), 'this gene')],\n",
            "[((111, 112), 'OA'), ((359, 360), 'OA'), ((364, 365), 'OA'), ((1394, 1395), 'OA')],\n",
            "[((372, 377), '310 disease - specific differentially expressed genes'), ((403, 406), 'these 310 genes'), ((639, 641), '117 genes'), ((654, 661), 'the 117 genes retrieved from Ensembl [ 30 ]')],\n",
            "[((429, 440), '16 genes , representing genes that were upregulated after induction of CIA'), ((443, 445), 'these genes'), ((461, 462), 'they')],\n",
            "[((303, 309), 'the onset phase of the disease'), ((449, 455), 'the onset phase of the disease')],\n",
            "[((316, 322), 'the chronic phase of the disease'), ((483, 488), 'the chronic phase of CIA'), ((544, 547), 'the chronic phase'), ((805, 810), 'the chronic phase of CIA'), ((890, 893), 'the chronic phase')],\n",
            "[((471, 488), '12 genes whose expression was gradually upregulated and reached a peak at the chronic phase of CIA'), ((488, 490), 'These genes')],\n",
            "[((507, 517), '78 genes that were only upregulated at the PI phase'), ((517, 519), 'These genes'), ((547, 549), 'These genes')],\n",
            "[((504, 506), 'Cluster III'), ((567, 569), 'Cluster V'), ((577, 579), 'Cluster VI')],\n",
            "[((514, 517), 'the PI phase'), ((574, 577), 'the PI phase'), ((584, 587), 'the OA phase'), ((779, 784), 'the OA phase of CIA'), ((790, 793), 'the PI phase'), ((1341, 1344), 'the OA phase')],\n",
            "[((607, 612), 'the small - effect QTL of CIA'), ((618, 622), 'the CIA small - effect QTL'), ((1224, 1227), 'the small - effect QTL')],\n",
            "[((664, 668), 'the 8 small - effect QTL'), ((668, 671), 'The eight loci'), ((707, 709), 'all loci'), ((720, 723), 'the eight QTL'), ((982, 986), 'the eight small - effect QTL'), ((1017, 1021), 'the eight small - effect QTL'), ((1030, 1032), 'these QTL'), ((1072, 1075), 'these eight QTL'), ((1082, 1086), 'the eight small - effect QTL'), ((1523, 1527), 'the eight small - effect QTL'), ((1731, 1745), 'the eight small - effect QTL identified using a very low threshold value ( P < 0 . 05 )')],\n",
            "[((775, 776), 'Oas1a'), ((785, 786), 'Oas1a'), ((813, 814), 'hspa1a')],\n",
            "[((771, 776), 'Two genes , hspa1a and Oas1a'), ((795, 798), 'these two genes')],\n",
            "[((339, 344), 'all three phases of CIA'), ((831, 835), 'all phases of CIA')],\n",
            "[((825, 827), 'Five genes'), ((843, 846), 'these five genes')],\n",
            "[((836, 837), 'H2 - Q10'), ((846, 847), 'H2 - Q10')],\n",
            "[((264, 267), 'the susceptible strain'), ((856, 859), 'the FVB / N strain'), ((1114, 1117), 'the resistant strain'), ((1354, 1356), 'the susceptible'), ((1431, 1434), 'the FVB / N strain'), ((1471, 1474), 'the FVB / N strain'), ((1495, 1498), 'the FVB / N strain')],\n",
            "[((100, 101), 'FVB / N'), ((857, 858), 'FVB / N'), ((1095, 1096), 'FVBN'), ((1384, 1385), 'FVB / N'), ((1432, 1433), 'FVB / N'), ((1496, 1497), 'FVB / N')],\n",
            "[((899, 906), 'the functional cluster of protein kinase cascade'), ((914, 917), 'This functional cluster')],\n",
            "[((837, 838), 'Mapk14'), ((921, 922), 'Mapk14')],\n",
            "[((769, 771), 'this study'), ((928, 930), 'this study'), ((1039, 1042), 'the present study')],\n",
            "[((691, 692), 'F2'), ((942, 943), 'F2'), ((1134, 1135), 'F2')],\n",
            "[((756, 760), 'loci 4 and 7'), ((1118, 1122), 'loci 2 and 7'), ((1545, 1549), 'QTL 4 and 7'), ((1558, 1561), 'these two QTL')],\n",
            "[((1147, 1157), 'Cia27 , a QTL controlling IgG2a antibody levels to collagen II'), ((1161, 1163), 'this QTL')],\n",
            "[((1164, 1169), 'a 4 . 1 Mb genomic region'), ((1175, 1177), 'this region')],\n",
            "[((1193, 1195), 'chromosome 16'), ((1211, 1213), 'chromosome 17')],\n",
            "[((1172, 1177), 'a gene within this region'), ((1221, 1227), 'the gene within the small - effect QTL')],\n",
            "[((168, 173), 'the onset and chronic phases'), ((1304, 1309), 'the PI and CA phases'), ((1334, 1339), 'the PI and CA phases')],\n",
            "[((1366, 1374), 'one cluster of genes ( cluster IV , Figure 2 )'), ((1374, 1376), 'These genes'), ((1409, 1411), 'these genes')],\n",
            "[((1512, 1514), 'Twenty - one genes'), ((1529, 1530), 'their'), ((1536, 1537), 'their')],\n",
            "[((1616, 1618), 'p38 MAPK'), ((1625, 1627), 'p38 MAPK')],\n",
            "[((1078, 1079), 'RA'), ((1634, 1635), 'RA'), ((1946, 1947), 'RA'), ((1958, 1959), 'RA'), ((2015, 2016), 'RA')],\n",
            "[((1636, 1643), 'Stat5a , a candidate gene for loci 6 ,'), ((1653, 1654), 'Stat5a'), ((1667, 1668), 'Stat5a')],\n",
            "[((1692, 1694), 'small - effect QTL'), ((1701, 1702), 'them')],\n",
            "[((1582, 1583), 'arthritis'), ((1768, 1769), 'arthritis'), ((1889, 1890), 'arthritis')],\n",
            "[((1835, 1844), 'a strategy to search candidate genes for small - effect QTL'), ((1845, 1847), 'this strategy')],\n",
            "[((1904, 1906), 'Candidate genes'), ((1911, 1912), 'their')],\n",
            "\n",
            "Mentions:\n",
            "((0, 2), 'We detected'),\n",
            "((0, 16), 'We detected the gene expression profiles using three mice per group , which is a small number'),\n",
            "((0, 16), 'We detected the gene expression profiles using three mice per group , which is a small number .'),\n",
            "((1, 2), 'detected'),\n",
            "((2, 6), 'the gene expression profiles'),\n",
            "((2, 16), 'the gene expression profiles using three mice per group , which is a small number'),\n",
            "((6, 7), 'using'),\n",
            "((7, 9), 'three mice'),\n",
            "((7, 11), 'three mice per group'),\n",
            "((7, 16), 'three mice per group , which is a small number'),\n",
            "((9, 11), 'per group'),\n",
            "((10, 11), 'group'),\n",
            "((12, 13), 'is'),\n",
            "((13, 16), 'a small number'),\n",
            "((16, 17), 'Being'),\n",
            "((16, 26), 'Being aware of the importance of data reproducibility , we determined'),\n",
            "((17, 26), 'aware of the importance of data reproducibility , we determined'),\n",
            "((19, 24), 'the importance of data reproducibility'),\n",
            "((19, 26), 'the importance of data reproducibility , we determined'),\n",
            "((22, 24), 'data reproducibility'),\n",
            "((23, 26), ', we determined'),\n",
            "((24, 26), 'we determined'),\n",
            "((25, 26), 'determined'),\n",
            "((26, 31), 'the coefficient of variation ( CV )'),\n",
            "((26, 35), 'the coefficient of variation ( CV ) to measure data variability'),\n",
            "((27, 31), 'coefficient of variation ( CV )'),\n",
            "((32, 33), 'measure'),\n",
            "((33, 35), 'data variability'),\n",
            "((35, 43), 'The CV for each gene on the chip'),\n",
            "((35, 52), 'The CV for each gene on the chip and the mean CV for the entire probe set'),\n",
            "((35, 54), 'The CV for each gene on the chip and the mean CV for the entire probe set were calculated'),\n",
            "((38, 43), 'each gene on the chip'),\n",
            "((41, 43), 'the chip'),\n",
            "((44, 47), 'the mean CV'),\n",
            "((44, 52), 'the mean CV for the entire probe set'),\n",
            "((48, 52), 'the entire probe set'),\n",
            "((52, 53), 'were'),\n",
            "((52, 54), 'were calculated'),\n",
            "((53, 54), 'calculated'),\n",
            "((54, 62), 'The mean CV ranged between 18 . 4 % and 25 . 8 %'),\n",
            "((54, 66), 'The mean CV ranged between 18 . 4 % and 25 . 8 % for all experimental groups'),\n",
            "((54, 66), 'The mean CV ranged between 18 . 4 % and 25 . 8 % for all experimental groups ,'),\n",
            "((54, 71), 'The mean CV ranged between 18 . 4 % and 25 . 8 % for all experimental groups , and this relatively low CV'),\n",
            "((57, 58), 'ranged'),\n",
            "((57, 62), 'ranged between 18 . 4 % and 25 . 8 %'),\n",
            "((63, 66), 'all experimental groups'),\n",
            "((73, 75), 'these data'),\n",
            "((73, 83), 'these data could be used for further analysis ( Table 1 ) .'),\n",
            "((79, 81), 'further analysis'),\n",
            "((81, 83), 'Table 1'),\n",
            "((81, 83), 'Table 1 )'),\n",
            "((81, 83), 'Table 1 ) .'),\n",
            "((82, 83), '.'),\n",
            "((84, 85), 'search'),\n",
            "((86, 90), 'strain - specific differentially expressed genes'),\n",
            "((91, 92), 'performed'),\n",
            "((94, 96), 'gene expression'),\n",
            "((98, 99), 'DB'),\n",
            "((98, 99), 'DBA'),\n",
            "((98, 114), 'DBA / 1 and FVB / N strains at all four phases of CIA , including NC , PI , OA and CA'),\n",
            "((98, 99), '/ 1'),\n",
            "((98, 99), '1'),\n",
            "((100, 102), 'FVB / N strains'),\n",
            "((100, 114), 'FVB / N strains at all four phases of CIA , including NC , PI , OA and CA'),\n",
            "((100, 101), 'N'),\n",
            "((103, 108), 'all four phases of CIA'),\n",
            "((109, 112), 'NC , PI , OA'),\n",
            "((109, 114), 'NC , PI , OA and CA'),\n",
            "((110, 112), 'PI , OA'),\n",
            "((110, 111), '##I'),\n",
            "((111, 112), 'O'),\n",
            "((115, 120), 'the naive mice without immunisation'),\n",
            "((115, 120), 'the naive mice without immunisation ,'),\n",
            "((115, 129), 'the naive mice without immunisation , 361 genes were differentially expressed between the two strains'),\n",
            "((119, 120), 'immunisation'),\n",
            "((120, 122), '361 genes'),\n",
            "((120, 129), '361 genes were differentially expressed between the two strains'),\n",
            "((123, 125), 'differentially expressed'),\n",
            "((129, 150), 'On day 10 after immunisation , when both strains did not show any sign of the disease , 141 genes were differentially expressed'),\n",
            "((130, 134), 'day 10 after immunisation'),\n",
            "((130, 150), 'day 10 after immunisation , when both strains did not show any sign of the disease , 141 genes were differentially expressed'),\n",
            "((133, 134), 'immunisation'),\n",
            "((135, 145), 'both strains did not show any sign of the disease'),\n",
            "((135, 145), 'both strains did not show any sign of the disease ,'),\n",
            "((135, 150), 'both strains did not show any sign of the disease , 141 genes were differentially expressed'),\n",
            "((137, 138), 'did'),\n",
            "((137, 140), 'did not show'),\n",
            "((138, 140), 'not show'),\n",
            "((139, 140), 'show'),\n",
            "((145, 147), '141 genes'),\n",
            "((145, 150), '141 genes were differentially expressed'),\n",
            "((148, 150), 'differentially expressed'),\n",
            "((149, 150), 'expressed'),\n",
            "((151, 153), 'DBA / 1 mice'),\n",
            "((151, 154), 'DBA / 1 mice developed'),\n",
            "((151, 155), 'DBA / 1 mice developed CIA'),\n",
            "((151, 155), 'DBA / 1 mice developed CIA ,'),\n",
            "((151, 174), 'DBA / 1 mice developed CIA , 184 and 85 differentially expressed genes were identified between these two strains at the onset and chronic phases , respectively'),\n",
            "((153, 154), 'developed'),\n",
            "((155, 161), '184 and 85 differentially expressed genes'),\n",
            "((155, 163), '184 and 85 differentially expressed genes were identified'),\n",
            "((155, 174), '184 and 85 differentially expressed genes were identified between these two strains at the onset and chronic phases , respectively'),\n",
            "((162, 163), 'identified'),\n",
            "((164, 174), 'these two strains at the onset and chronic phases , respectively'),\n",
            "((168, 174), 'the onset and chronic phases , respectively'),\n",
            "((171, 172), 'chronic'),\n",
            "((173, 174), 'respectively'),\n",
            "((175, 186), 'the lists of the differentially expressed genes at the four phases'),\n",
            "((175, 197), 'the lists of the differentially expressed genes at the four phases were merged and overlapping genes were excluded , 509 genes were identified'),\n",
            "((178, 186), 'the differentially expressed genes at the four phases'),\n",
            "((187, 188), 'merged'),\n",
            "((189, 191), 'overlapping genes'),\n",
            "((192, 193), 'excluded'),\n",
            "((193, 195), '509 genes'),\n",
            "((193, 197), '509 genes were identified'),\n",
            "((196, 197), 'identified'),\n",
            "((197, 200), 'Additional file 1'),\n",
            "((197, 200), 'Additional file 1 )'),\n",
            "((200, 210), 'Twenty - one genes consistently showed differential expression between the two strains'),\n",
            "((200, 213), 'Twenty - one genes consistently showed differential expression between the two strains at all phases'),\n",
            "((202, 204), 'consistently showed'),\n",
            "((203, 204), 'showed'),\n",
            "((217, 221), 'only 3 additional genes'),\n",
            "((217, 232), 'only 3 additional genes were strain - specific differentially expressed during the 3 phases after CIA induction'),\n",
            "((217, 238), 'only 3 additional genes were strain - specific differentially expressed during the 3 phases after CIA induction ( PI , OA and CA ; Figure 1 )'),\n",
            "((221, 222), 'were'),\n",
            "((222, 225), 'strain - specific differentially expressed'),\n",
            "((226, 232), 'the 3 phases after CIA induction'),\n",
            "((230, 232), 'CIA induction'),\n",
            "((232, 236), 'PI , OA and CA'),\n",
            "((232, 238), 'PI , OA and CA ; Figure 1 )'),\n",
            "((232, 233), '##I'),\n",
            "((233, 234), 'OA'),\n",
            "((236, 238), 'Figure 1'),\n",
            "((236, 238), 'Figure 1 )'),\n",
            "((237, 238), '1'),\n",
            "((238, 242), 'Disease - specific differentially expressed genes'),\n",
            "((242, 244), 'To identify'),\n",
            "((243, 244), 'identify'),\n",
            "((244, 251), 'the disease - specific differentially expressed genes in CIA'),\n",
            "((244, 267), 'the disease - specific differentially expressed genes in CIA , we detected the genes that were differentially expressed in LNs during CIA in the susceptible strain'),\n",
            "((251, 253), 'we detected'),\n",
            "((251, 267), 'we detected the genes that were differentially expressed in LNs during CIA in the susceptible strain'),\n",
            "((252, 253), 'detected'),\n",
            "((253, 263), 'the genes that were differentially expressed in LNs during CIA'),\n",
            "((253, 267), 'the genes that were differentially expressed in LNs during CIA in the susceptible strain'),\n",
            "((257, 259), 'differentially expressed'),\n",
            "((258, 259), 'expressed'),\n",
            "((260, 261), 'LN'),\n",
            "((260, 261), 'LNs'),\n",
            "((260, 263), 'LNs during CIA'),\n",
            "((265, 266), 'susceptible'),\n",
            "((265, 267), 'susceptible strain'),\n",
            "((266, 267), 'strain'),\n",
            "((267, 270), 'Three experimental conditions'),\n",
            "((267, 274), 'Three experimental conditions , PI , OA and CA'),\n",
            "((267, 274), 'Three experimental conditions , PI , OA and CA ,'),\n",
            "((267, 280), 'Three experimental conditions , PI , OA and CA , were compared with the NC group . [SEP]'),\n",
            "((270, 274), 'PI , OA and CA'),\n",
            "((271, 272), 'OA'),\n",
            "((274, 275), 'were'),\n",
            "((275, 276), 'compared'),\n",
            "((275, 280), 'compared with the NC group . [SEP]'),\n",
            "((277, 280), 'the NC group'),\n",
            "((277, 280), 'the NC group .'),\n",
            "((277, 280), 'the NC group . [SEP]'),\n",
            "((278, 280), 'NC group'),\n",
            "((279, 280), 'group'),\n",
            "((280, 285), 'On day 10 after immunisation'),\n",
            "((280, 300), 'On day 10 after immunisation , 102 genes were differentially expressed – most of them were upregulated ( 78 out of 102 )'),\n",
            "((281, 285), 'day 10 after immunisation'),\n",
            "((284, 285), 'immunisation'),\n",
            "((285, 300), '102 genes were differentially expressed – most of them were upregulated ( 78 out of 102 )'),\n",
            "((287, 288), 'were'),\n",
            "((288, 290), 'differentially expressed'),\n",
            "((296, 300), '78 out of 102'),\n",
            "((299, 300), '102'),\n",
            "((303, 315), 'the onset phase of the disease , only 26 genes were differentially expressed'),\n",
            "((309, 312), 'only 26 genes'),\n",
            "((309, 315), 'only 26 genes were differentially expressed'),\n",
            "((313, 315), 'differentially expressed'),\n",
            "((316, 332), 'the chronic phase of the disease , 184 differentially expressed genes were identified , with 156 downregulated genes'),\n",
            "((322, 326), '184 differentially expressed genes'),\n",
            "((322, 328), '184 differentially expressed genes were identified'),\n",
            "((322, 332), '184 differentially expressed genes were identified , with 156 downregulated genes'),\n",
            "((327, 328), 'identified'),\n",
            "((329, 330), '156'),\n",
            "((329, 332), '156 downregulated genes'),\n",
            "((332, 344), 'Only one gene was differentially expressed at all three phases of CIA'),\n",
            "((336, 338), 'differentially expressed'),\n",
            "((345, 354), 'this gene , five , one and six differentially expressed genes'),\n",
            "((347, 354), 'five , one and six differentially expressed genes'),\n",
            "((357, 360), 'PI with OA'),\n",
            "((357, 367), 'PI with OA , PI with CA and OA with CA'),\n",
            "((357, 367), 'PI with OA , PI with CA and OA with CA ,'),\n",
            "((357, 368), 'PI with OA , PI with CA and OA with CA , respectively'),\n",
            "((357, 370), 'PI with OA , PI with CA and OA with CA , respectively ( Figure 2a )'),\n",
            "((357, 370), 'PI with OA , PI with CA and OA with CA , respectively ( Figure 2a ) .'),\n",
            "((360, 367), 'PI with CA and OA with CA'),\n",
            "((362, 367), 'CA and OA with CA'),\n",
            "((364, 367), 'OA with CA'),\n",
            "((368, 370), 'Figure 2a'),\n",
            "((368, 370), 'Figure 2a )'),\n",
            "((372, 382), '310 disease - specific differentially expressed genes were differentially regulated during CIA'),\n",
            "((372, 385), '310 disease - specific differentially expressed genes were differentially regulated during CIA in DBA / 1 mice'),\n",
            "((377, 378), 'were'),\n",
            "((378, 380), 'differentially regulated'),\n",
            "((383, 384), 'DBA / 1'),\n",
            "((383, 385), 'DBA / 1 mice'),\n",
            "((383, 384), '/ 1'),\n",
            "((383, 384), '1'),\n",
            "((385, 388), 'Additional file 2'),\n",
            "((385, 388), 'Additional file 2 )'),\n",
            "((385, 388), 'Additional file 2 ) .'),\n",
            "((391, 397), 'the gene expression pattern during CIA'),\n",
            "((391, 406), 'the gene expression pattern during CIA , we performed hierarchical cluster analysis for these 310 genes'),\n",
            "((397, 399), 'we performed'),\n",
            "((398, 399), 'performed'),\n",
            "((399, 406), 'hierarchical cluster analysis for these 310 genes'),\n",
            "((406, 409), 'Six gene clusters'),\n",
            "((406, 426), 'Six gene clusters were identified ( clusters I to VI , Figure 2b ) , each with a distinct gene expression pattern during CIA'),\n",
            "((409, 411), 'were identified'),\n",
            "((410, 411), 'identified'),\n",
            "((411, 415), 'clusters I to VI'),\n",
            "((411, 417), 'clusters I to VI , Figure 2b'),\n",
            "((411, 417), 'clusters I to VI , Figure 2b )'),\n",
            "((411, 426), 'clusters I to VI , Figure 2b ) , each with a distinct gene expression pattern during CIA'),\n",
            "((412, 413), 'I'),\n",
            "((414, 415), 'VI'),\n",
            "((415, 417), 'Figure 2b'),\n",
            "((417, 426), 'each with a distinct gene expression pattern during CIA'),\n",
            "((419, 426), 'a distinct gene expression pattern during CIA'),\n",
            "((426, 428), 'Cluster I'),\n",
            "((426, 440), 'Cluster I contains 16 genes , representing genes that were upregulated after induction of CIA'),\n",
            "((428, 429), 'contains'),\n",
            "((432, 440), 'genes that were upregulated after induction of CIA'),\n",
            "((437, 440), 'induction of CIA'),\n",
            "((440, 445), 'The expression of these genes'),\n",
            "((440, 455), 'The expression of these genes reached a peak at the onset phase of the disease'),\n",
            "((440, 468), 'The expression of these genes reached a peak at the onset phase of the disease and functional clustering results revealed that they are related to the immune response'),\n",
            "((445, 446), 'reached'),\n",
            "((465, 468), 'the immune response'),\n",
            "((468, 470), 'Cluster II'),\n",
            "((468, 488), 'Cluster II contains 12 genes whose expression was gradually upregulated and reached a peak at the chronic phase of CIA'),\n",
            "((470, 471), 'contains'),\n",
            "((488, 504), 'These genes are mainly related to the immune response , organelle membrane and extracellular region and space'),\n",
            "((494, 497), 'the immune response'),\n",
            "((494, 504), 'the immune response , organelle membrane and extracellular region and space'),\n",
            "((497, 499), 'organelle membrane'),\n",
            "((497, 504), 'organelle membrane and extracellular region and space'),\n",
            "((500, 504), 'extracellular region and space'),\n",
            "((504, 517), 'Cluster III contains 78 genes that were only upregulated at the PI phase'),\n",
            "((506, 507), 'contains'),\n",
            "((515, 516), 'PI'),\n",
            "((517, 525), 'These genes are related to the intercellular junction'),\n",
            "((522, 525), 'the intercellular junction'),\n",
            "((525, 534), 'More than half of the genes ( 156 of 310 )'),\n",
            "((525, 535), 'More than half of the genes ( 156 of 310 ) belong'),\n",
            "((525, 538), 'More than half of the genes ( 156 of 310 ) belong to cluster IV'),\n",
            "((525, 540), 'More than half of the genes ( 156 of 310 ) belong to cluster IV and represent'),\n",
            "((525, 547), 'More than half of the genes ( 156 of 310 ) belong to cluster IV and represent genes specifically downregultaed at the chronic phase . [SEP]'),\n",
            "((529, 531), 'the genes'),\n",
            "((529, 534), 'the genes ( 156 of 310 )'),\n",
            "((531, 534), '156 of 310'),\n",
            "((531, 534), '156 of 310 )'),\n",
            "((533, 534), '310'),\n",
            "((534, 535), 'belong'),\n",
            "((536, 538), 'cluster IV'),\n",
            "((539, 540), 'represent'),\n",
            "((540, 547), 'genes specifically downregultaed at the chronic phase . [SEP]'),\n",
            "((544, 547), 'the chronic phase .'),\n",
            "((544, 547), 'the chronic phase . [SEP]'),\n",
            "((547, 567), 'These genes are functionally related to lymphocyte proliferation , T cell activation , protein binding as well as the notch signal pathway'),\n",
            "((553, 555), 'lymphocyte proliferation'),\n",
            "((553, 560), 'lymphocyte proliferation , T cell activation , protein binding'),\n",
            "((553, 567), 'lymphocyte proliferation , T cell activation , protein binding as well as the notch signal pathway'),\n",
            "((555, 557), 'T cell'),\n",
            "((555, 558), 'T cell activation'),\n",
            "((558, 560), 'protein binding'),\n",
            "((563, 567), 'the notch signal pathway'),\n",
            "((567, 577), 'Cluster V contains eight genes downregulated at the PI phase'),\n",
            "((569, 570), 'contains'),\n",
            "((570, 573), 'eight genes downregulated'),\n",
            "((570, 577), 'eight genes downregulated at the PI phase'),\n",
            "((575, 576), 'PI'),\n",
            "((577, 587), 'Cluster VI contains 18 genes downregulated at the OA phase'),\n",
            "((579, 580), 'contains'),\n",
            "((580, 587), '18 genes downregulated at the OA phase'),\n",
            "((585, 586), 'OA'),\n",
            "((587, 591), 'The GO term classification'),\n",
            "((587, 604), 'The GO term classification showed no functional cluster that was significantly enriched in these two gene clusters'),\n",
            "((591, 592), 'showed'),\n",
            "((604, 612), 'Candidate genes for the small - effect QTL of CIA'),\n",
            "((612, 624), 'To identify candidate susceptibility genes for the CIA small - effect QTL , we compared'),\n",
            "((613, 614), 'identify'),\n",
            "((613, 624), 'identify candidate susceptibility genes for the CIA small - effect QTL , we compared'),\n",
            "((614, 622), 'candidate susceptibility genes for the CIA small - effect QTL'),\n",
            "((614, 622), 'candidate susceptibility genes for the CIA small - effect QTL ,'),\n",
            "((614, 624), 'candidate susceptibility genes for the CIA small - effect QTL , we compared'),\n",
            "((622, 624), 'we compared'),\n",
            "((623, 624), 'compared'),\n",
            "((624, 631), 'the list of strain - specific differentially expressed genes'),\n",
            "((624, 639), 'the list of strain - specific differentially expressed genes with the list of disease - specific differentially expressed genes'),\n",
            "((624, 641), 'the list of strain - specific differentially expressed genes with the list of disease - specific differentially expressed genes ; 117 genes'),\n",
            "((624, 646), 'the list of strain - specific differentially expressed genes with the list of disease - specific differentially expressed genes ; 117 genes were shared by both lists'),\n",
            "((627, 630), 'strain - specific differentially expressed'),\n",
            "((632, 639), 'the list of disease - specific differentially expressed genes'),\n",
            "((635, 638), 'disease - specific differentially expressed'),\n",
            "((635, 639), 'disease - specific differentially expressed genes'),\n",
            "((646, 649), 'Additional file 3 )'),\n",
            "((649, 651), 'Figure 3'),\n",
            "((649, 668), 'Figure 3 visualises positions of the 117 genes retrieved from Ensembl [ 30 ] in relation to the 8 small - effect QTL'),\n",
            "((651, 652), 'visualises'),\n",
            "((654, 668), 'the 117 genes retrieved from Ensembl [ 30 ] in relation to the 8 small - effect QTL'),\n",
            "((659, 661), 'Ensembl [ 30 ]'),\n",
            "((668, 684), 'The eight loci were located on 7 chromosomes , 5 , 6 , 7 , 10 , 11 , 16 and 17'),\n",
            "((672, 673), 'located'),\n",
            "((672, 684), 'located on 7 chromosomes , 5 , 6 , 7 , 10 , 11 , 16 and 17'),\n",
            "((674, 684), '7 chromosomes , 5 , 6 , 7 , 10 , 11 , 16 and 17'),\n",
            "((676, 677), '5'),\n",
            "((679, 680), '10'),\n",
            "((680, 681), '11'),\n",
            "((681, 682), '16'),\n",
            "((683, 684), '17'),\n",
            "((685, 693), 'the confidence intervals of QTL in F2 progeny'),\n",
            "((685, 698), 'the confidence intervals of QTL in F2 progeny are around 20 cM [ 26 ]'),\n",
            "((685, 702), 'the confidence intervals of QTL in F2 progeny are around 20 cM [ 26 ] , we used 40 Mb'),\n",
            "((689, 690), 'QTL'),\n",
            "((689, 693), 'QTL in F2 progeny'),\n",
            "((691, 693), 'F2 pro'),\n",
            "((691, 693), 'F2 progeny'),\n",
            "((693, 694), 'are'),\n",
            "((698, 700), 'we used'),\n",
            "((698, 702), 'we used 40 Mb'),\n",
            "((699, 700), 'used'),\n",
            "((700, 702), '40 Mb'),\n",
            "((703, 709), 'the confidence intervals for all loci'),\n",
            "((709, 711), 'Twenty - one genes'),\n",
            "((709, 723), 'Twenty - one genes were located in the confidence intervals of six of the eight QTL'),\n",
            "((712, 713), 'located'),\n",
            "((714, 723), 'the confidence intervals of six of the eight QTL'),\n",
            "((718, 723), 'six of the eight QTL'),\n",
            "((723, 725), 'We located'),\n",
            "((723, 735), 'We located 5 , 4 , 2 , 1 , 3 and 6 potential candidate genes'),\n",
            "((724, 725), 'located'),\n",
            "((725, 726), '5'),\n",
            "((725, 732), '5 , 4 , 2 , 1 , 3 and 6'),\n",
            "((725, 735), '5 , 4 , 2 , 1 , 3 and 6 potential candidate genes'),\n",
            "((731, 732), '6'),\n",
            "((736, 748), 'the confidence intervals of loci 1 , 2 , 3 , 5 , 6 and 8'),\n",
            "((736, 749), 'the confidence intervals of loci 1 , 2 , 3 , 5 , 6 and 8 , respectively'),\n",
            "((736, 749), 'the confidence intervals of loci 1 , 2 , 3 , 5 , 6 and 8 , respectively ,'),\n",
            "((740, 748), 'loci 1 , 2 , 3 , 5 , 6 and 8'),\n",
            "((744, 745), '5'),\n",
            "((745, 746), '6'),\n",
            "((747, 748), '8'),\n",
            "((748, 749), 'respectively'),\n",
            "((754, 755), 'identified'),\n",
            "((756, 760), 'loci 4 and 7 .'),\n",
            "((760, 762), 'Table 3'),\n",
            "((760, 771), 'Table 3 summarises the 21 candidate genes identified in this study'),\n",
            "((760, 771), 'Table 3 summarises the 21 candidate genes identified in this study .'),\n",
            "((762, 763), 'summarises'),\n",
            "((763, 771), 'the 21 candidate genes identified in this study'),\n",
            "((767, 768), 'identified'),\n",
            "((771, 776), 'Two genes , hspa1a and Oas1a ,'),\n",
            "((773, 774), 'hspa1a'),\n",
            "((773, 776), 'hspa1a and Oas1a'),\n",
            "((775, 776), ','),\n",
            "((776, 777), 'were'),\n",
            "((779, 793), 'the OA phase of CIA and Oas1a was also upregulated at the PI phase'),\n",
            "((779, 793), 'the OA phase of CIA and Oas1a was also upregulated at the PI phase .'),\n",
            "((779, 793), 'the OA phase of CIA and Oas1a was also upregulated at the PI phase . [SEP]'),\n",
            "((780, 781), 'OA'),\n",
            "((785, 793), 'Oas1a was also upregulated at the PI phase'),\n",
            "((791, 792), 'PI'),\n",
            "((795, 802), 'these two genes , all other 19 genes'),\n",
            "((795, 810), 'these two genes , all other 19 genes were downregulated at the chronic phase of CIA'),\n",
            "((798, 802), 'all other 19 genes'),\n",
            "((810, 812), 'All genes'),\n",
            "((810, 814), 'All genes , except hspa1a'),\n",
            "((810, 814), 'All genes , except hspa1a ,'),\n",
            "((810, 821), 'All genes , except hspa1a , showed expression differences between the two strains'),\n",
            "((810, 825), 'All genes , except hspa1a , showed expression differences between the two strains at the NC phase'),\n",
            "((810, 825), 'All genes , except hspa1a , showed expression differences between the two strains at the NC phase .'),\n",
            "((814, 815), 'showed'),\n",
            "((822, 825), 'the NC phase'),\n",
            "((828, 830), 'differentially expressed'),\n",
            "((829, 830), 'expressed'),\n",
            "((836, 842), 'H2 - Q10 , Mapk14 , Pscd1 , Kpnb1 and Wdr1'),\n",
            "((837, 842), 'Mapk14 , Pscd1 , Kpnb1 and Wdr1'),\n",
            "((838, 839), 'Pscd'),\n",
            "((838, 839), 'Pscd1'),\n",
            "((838, 840), 'Pscd1 , Kpnb1'),\n",
            "((838, 842), 'Pscd1 , Kpnb1 and Wdr1'),\n",
            "((839, 840), 'Kpnb1'),\n",
            "((841, 842), 'Wdr1'),\n",
            "((846, 863), 'H2 - Q10 had a consistently higher expression in the DBA / 1 than the FVB / N strain in all CIA phases'),\n",
            "((846, 863), 'H2 - Q10 had a consistently higher expression in the DBA / 1 than the FVB / N strain in all CIA phases ,'),\n",
            "((847, 848), 'had'),\n",
            "((848, 863), 'a consistently higher expression in the DBA / 1 than the FVB / N strain in all CIA phases'),\n",
            "((853, 863), 'the DBA / 1 than the FVB / N strain in all CIA phases'),\n",
            "((856, 863), 'the FVB / N strain in all CIA phases'),\n",
            "((857, 858), 'F'),\n",
            "((857, 858), 'FVB'),\n",
            "((857, 859), 'FVB / N strain'),\n",
            "((860, 863), 'all CIA phases'),\n",
            "((864, 868), 'the other four genes'),\n",
            "((868, 869), 'had'),\n",
            "((873, 885), 'the DBA / 1 strain at the early stages , including NC , PI and OA'),\n",
            "((873, 885), 'the DBA / 1 strain at the early stages , including NC , PI and OA ,'),\n",
            "((873, 893), 'the DBA / 1 strain at the early stages , including NC , PI and OA , but a lower expression at the chronic phase'),\n",
            "((877, 885), 'the early stages , including NC , PI and OA'),\n",
            "((881, 885), 'NC , PI and OA'),\n",
            "((882, 885), 'PI and OA'),\n",
            "((882, 883), '##I'),\n",
            "((884, 885), 'O'),\n",
            "((884, 885), 'OA'),\n",
            "((893, 894), 'GO'),\n",
            "((893, 897), 'GO term classification analysis'),\n",
            "((893, 914), 'GO term classification analysis revealed that the functional cluster of protein kinase cascade was significantly enriched in the 21 candidate genes'),\n",
            "((897, 898), 'revealed'),\n",
            "((899, 914), 'the functional cluster of protein kinase cascade was significantly enriched in the 21 candidate genes'),\n",
            "((903, 906), 'protein kinase cascade'),\n",
            "((906, 907), 'was'),\n",
            "((910, 914), 'the 21 candidate genes'),\n",
            "((914, 926), 'This functional cluster contained four genes , including Mapk14 , Mapk8ip3 , Stat5a and Gna12'),\n",
            "((917, 918), 'contained'),\n",
            "((918, 926), 'four genes , including Mapk14 , Mapk8ip3 , Stat5a and Gna12'),\n",
            "((921, 926), 'Mapk14 , Mapk8ip3 , Stat5a and Gna12'),\n",
            "((922, 923), 'Mapk8ip3'),\n",
            "((922, 926), 'Mapk8ip3 , Stat5a and Gna12'),\n",
            "((923, 924), 'Stat5'),\n",
            "((923, 924), 'Stat5a'),\n",
            "((923, 926), 'Stat5a and Gna12'),\n",
            "((925, 926), 'Gna'),\n",
            "((925, 926), 'Gna12'),\n",
            "((926, 927), 'Discussion'),\n",
            "((928, 944), 'this study , we attempted for the first time to identify small - effect QTL in an F2 progeny'),\n",
            "((928, 944), 'this study , we attempted for the first time to identify small - effect QTL in an F2 progeny .'),\n",
            "((929, 930), 'study'),\n",
            "((930, 938), 'we attempted for the first time to identify'),\n",
            "((930, 944), 'we attempted for the first time to identify small - effect QTL in an F2 progeny'),\n",
            "((931, 932), 'attempted'),\n",
            "((931, 936), 'attempted for the first time'),\n",
            "((931, 938), 'attempted for the first time to identify'),\n",
            "((933, 936), 'the first time'),\n",
            "((937, 938), 'identify'),\n",
            "((938, 944), 'small - effect QTL in an F2 progeny'),\n",
            "((941, 944), 'an F2 progeny'),\n",
            "((942, 944), 'F2 pro'),\n",
            "((942, 943), '##2'),\n",
            "((944, 946), 'Small - effect QTL'),\n",
            "((947, 948), 'defined'),\n",
            "((951, 958), 'the threshold value of P = 0 . 05'),\n",
            "((951, 973), 'the threshold value of P = 0 . 05 but that did not reach the significant threshold value suggested by Lander and Botstein [ 26 ]'),\n",
            "((955, 958), 'P = 0 . 05'),\n",
            "((969, 972), 'Lander and Botstein'),\n",
            "((969, 973), 'Lander and Botstein [ 26 ]'),\n",
            "((971, 972), 'Botstein'),\n",
            "((972, 973), '[ 26 ]'),\n",
            "((980, 986), 'most of the eight small - effect QTL'),\n",
            "((980, 988), 'most of the eight small - effect QTL likely contain'),\n",
            "((980, 992), 'most of the eight small - effect QTL likely contain susceptibility genes for CIA'),\n",
            "((980, 992), 'most of the eight small - effect QTL likely contain susceptibility genes for CIA .'),\n",
            "((986, 987), 'likely'),\n",
            "((986, 988), 'likely contain'),\n",
            "((987, 988), 'contain'),\n",
            "((988, 992), 'susceptibility genes for CIA'),\n",
            "((992, 993), 'First'),\n",
            "((992, 995), 'First , we performed'),\n",
            "((993, 995), 'we performed'),\n",
            "((994, 995), 'performed'),\n",
            "((995, 998), 'the linkage analysis'),\n",
            "((1002, 1006), 'all 290 F progeny'),\n",
            "((1002, 1015), 'all 290 F progeny , 76 C5 + / + F2 progeny and 133 C5 + / - F2 progeny'),\n",
            "((1004, 1005), 'F'),\n",
            "((1006, 1010), '76 C5 + / + F2 progeny'),\n",
            "((1006, 1015), '76 C5 + / + F2 progeny and 133 C5 + / - F2 progeny'),\n",
            "((1008, 1009), 'F'),\n",
            "((1008, 1009), 'F2'),\n",
            "((1008, 1010), 'F2 pro'),\n",
            "((1008, 1010), 'F2 progeny'),\n",
            "((1011, 1014), '133 C5 + / - F2'),\n",
            "((1011, 1015), '133 C5 + / - F2 pro'),\n",
            "((1011, 1015), '133 C5 + / - F2 progeny'),\n",
            "((1012, 1014), 'C5 + / - F2'),\n",
            "((1012, 1014), '+ / - F2'),\n",
            "((1012, 1014), '/ - F2'),\n",
            "((1012, 1014), '- F2'),\n",
            "((1013, 1014), 'F'),\n",
            "((1013, 1014), 'F2'),\n",
            "((1014, 1015), '.'),\n",
            "((1015, 1021), 'Five of the eight small - effect QTL'),\n",
            "((1015, 1023), 'Five of the eight small - effect QTL were identified'),\n",
            "((1015, 1028), 'Five of the eight small - effect QTL were identified in at least two datasets'),\n",
            "((1015, 1032), 'Five of the eight small - effect QTL were identified in at least two datasets , suggesting that these QTL'),\n",
            "((1015, 1034), 'Five of the eight small - effect QTL were identified in at least two datasets , suggesting that these QTL are reproducible'),\n",
            "((1015, 1034), 'Five of the eight small - effect QTL were identified in at least two datasets , suggesting that these QTL are reproducible .'),\n",
            "((1021, 1022), 'were'),\n",
            "((1021, 1023), 'were identified'),\n",
            "((1022, 1023), 'identified'),\n",
            "((1024, 1028), 'at least two datasets'),\n",
            "((1033, 1034), 'reproducible'),\n",
            "((1034, 1035), 'Second'),\n",
            "((1035, 1042), 'many QTL identified in the present study'),\n",
            "((1035, 1055), 'many QTL identified in the present study overlap with arthritis QTL previously identified , including loci 1 , 2 , 5 and 6'),\n",
            "((1035, 1055), 'many QTL identified in the present study overlap with arthritis QTL previously identified , including loci 1 , 2 , 5 and 6 .'),\n",
            "((1037, 1042), 'identified in the present study'),\n",
            "((1042, 1043), 'overlap'),\n",
            "((1044, 1048), 'arthritis QTL previously identified'),\n",
            "((1044, 1055), 'arthritis QTL previously identified , including loci 1 , 2 , 5 and 6'),\n",
            "((1046, 1048), 'previously identified'),\n",
            "((1047, 1048), 'identified'),\n",
            "((1048, 1055), 'including loci 1 , 2 , 5 and 6'),\n",
            "((1049, 1055), 'loci 1 , 2 , 5 and 6'),\n",
            "((1050, 1051), '1'),\n",
            "((1051, 1052), '2'),\n",
            "((1052, 1053), '5'),\n",
            "((1054, 1055), '6'),\n",
            "((1057, 1059), 'syntenic analysis'),\n",
            "((1057, 1075), 'syntenic analysis revealed that the counterpart genomic regions on the human genome of many of these eight QTL'),\n",
            "((1059, 1060), 'revealed'),\n",
            "((1061, 1075), 'the counterpart genomic regions on the human genome of many of these eight QTL'),\n",
            "((1066, 1069), 'the human genome'),\n",
            "((1070, 1075), 'many of these eight QTL'),\n",
            "((1076, 1077), 'linked'),\n",
            "((1080, 1086), 'five of the eight small - effect QTL'),\n",
            "((1080, 1089), 'five of the eight small - effect QTL the DBA / 1 alleles'),\n",
            "((1080, 1093), 'five of the eight small - effect QTL the DBA / 1 alleles are the arthritis - enhancing alleles'),\n",
            "((1086, 1089), 'the DBA / 1 alleles'),\n",
            "((1087, 1089), 'DBA / 1 allele'),\n",
            "((1087, 1089), 'DBA / 1 alleles'),\n",
            "((1087, 1088), '1'),\n",
            "((1089, 1090), 'are'),\n",
            "((1090, 1093), 'the arthritis - enhancing alleles'),\n",
            "((1091, 1092), 'arthritis - enhancing'),\n",
            "((1094, 1097), 'the FVBN alleles'),\n",
            "((1094, 1106), 'the FVBN alleles are the arthritis - enhancing alleles in the other three QTL'),\n",
            "((1095, 1097), 'FVBN alleles'),\n",
            "((1097, 1098), 'are'),\n",
            "((1098, 1106), 'the arthritis - enhancing alleles in the other three QTL'),\n",
            "((1102, 1106), 'the other three QTL'),\n",
            "((1108, 1111), 'some susceptibility genes'),\n",
            "((1108, 1113), 'some susceptibility genes could come'),\n",
            "((1109, 1111), 'susceptibility genes'),\n",
            "((1112, 1113), 'come'),\n",
            "((1117, 1118), 'Interestingly'),\n",
            "((1118, 1120), 'loci 2'),\n",
            "((1118, 1137), 'loci 2 and 7 partially overlap with two arthritis - related QTL identified by us in the same F2 progeny [ 10 ]'),\n",
            "((1119, 1120), '2'),\n",
            "((1121, 1122), '7'),\n",
            "((1123, 1124), 'overlap'),\n",
            "((1125, 1137), 'two arthritis - related QTL identified by us in the same F2 progeny [ 10 ]'),\n",
            "((1130, 1137), 'us in the same F2 progeny [ 10 ]'),\n",
            "((1132, 1136), 'the same F2 pro'),\n",
            "((1132, 1136), 'the same F2 progeny'),\n",
            "((1132, 1137), 'the same F2 progeny [ 10 ]'),\n",
            "((1134, 1135), 'F'),\n",
            "((1134, 1136), 'F2 pro'),\n",
            "((1136, 1137), '[ 10'),\n",
            "((1136, 1137), '[ 10 ]'),\n",
            "((1136, 1137), '10'),\n",
            "((1137, 1139), 'Locus 2'),\n",
            "((1139, 1141), 'was located'),\n",
            "((1140, 1141), 'located'),\n",
            "((1142, 1157), 'the same genomic region as Cia27 , a QTL controlling IgG2a antibody levels to collagen II'),\n",
            "((1148, 1157), 'a QTL controlling IgG2a antibody levels to collagen II'),\n",
            "((1151, 1152), 'IgG2a'),\n",
            "((1151, 1153), 'IgG2a antibody'),\n",
            "((1151, 1154), 'IgG2a antibody levels'),\n",
            "((1151, 1155), 'IgG2a antibody levels to'),\n",
            "((1151, 1157), 'IgG2a antibody levels to collagen II'),\n",
            "((1155, 1157), 'collagen II'),\n",
            "((1157, 1158), 'Recently'),\n",
            "((1158, 1169), 'we have refined this QTL into a 4 . 1 Mb genomic region'),\n",
            "((1161, 1169), 'this QTL into a 4 . 1 Mb genomic region'),\n",
            "((1170, 1171), 'showed'),\n",
            "((1172, 1180), 'a gene within this region regulates CIA severity'),\n",
            "((1172, 1190), 'a gene within this region regulates CIA severity by controlling the IgG2a antibody levels to collagen II [ 31 ]'),\n",
            "((1177, 1178), 'regulate'),\n",
            "((1177, 1178), 'regulates'),\n",
            "((1178, 1180), 'CIA severity'),\n",
            "((1181, 1182), 'controlling'),\n",
            "((1182, 1189), 'the IgG2a antibody levels to collagen II'),\n",
            "((1182, 1190), 'the IgG2a antibody levels to collagen II [ 31 ]'),\n",
            "((1183, 1184), 'IgG2a'),\n",
            "((1183, 1185), 'IgG2a antibody'),\n",
            "((1183, 1189), 'IgG2a antibody levels to collagen II'),\n",
            "((1187, 1189), 'collagen II'),\n",
            "((1189, 1190), '[ 31 ]'),\n",
            "((1190, 1195), 'Locus 7 on chromosome 16'),\n",
            "((1190, 1202), 'Locus 7 on chromosome 16 overlaps with Lp1 , which controls lymphocyte proliferation'),\n",
            "((1195, 1196), 'overlap'),\n",
            "((1195, 1196), 'overlaps'),\n",
            "((1197, 1202), 'Lp1 , which controls lymphocyte proliferation'),\n",
            "((1199, 1200), 'controls'),\n",
            "((1200, 1202), 'lymphocyte proliferation'),\n",
            "((1202, 1203), 'Furthermore'),\n",
            "((1205, 1208), 'our unpublished data'),\n",
            "((1205, 1220), 'our unpublished data , loci 8 on chromosome 17 controls lymphocyte adherence during development of CIA'),\n",
            "((1208, 1213), 'loci 8 on chromosome 17'),\n",
            "((1208, 1214), 'loci 8 on chromosome 17 controls'),\n",
            "((1208, 1220), 'loci 8 on chromosome 17 controls lymphocyte adherence during development of CIA'),\n",
            "((1209, 1213), '8 on chromosome 17'),\n",
            "((1213, 1214), 'controls'),\n",
            "((1214, 1216), 'lymphocyte adherence'),\n",
            "((1214, 1220), 'lymphocyte adherence during development of CIA'),\n",
            "((1215, 1216), 'adherence'),\n",
            "((1217, 1218), 'development'),\n",
            "((1217, 1220), 'development of CIA'),\n",
            "((1220, 1221), 'Therefore'),\n",
            "((1228, 1229), 'affect'),\n",
            "((1229, 1231), 'CIA severity'),\n",
            "((1232, 1233), 'controlling'),\n",
            "((1232, 1235), 'controlling arthritis - related phenotypes'),\n",
            "((1233, 1234), 'arthritis - related'),\n",
            "((1233, 1235), 'arthritis - related phenotypes'),\n",
            "((1233, 1234), 'related'),\n",
            "((1235, 1237), 'Several studies'),\n",
            "((1239, 1240), 'carried'),\n",
            "((1239, 1241), 'carried out'),\n",
            "((1239, 1257), 'carried out to detect gene expression during CIA , all of which used joints as the target tissue [ 15 , 16 , 21 , 22 ]'),\n",
            "((1242, 1243), 'detect'),\n",
            "((1242, 1245), 'detect gene expression'),\n",
            "((1243, 1245), 'gene expression'),\n",
            "((1246, 1257), 'CIA , all of which used joints as the target tissue [ 15 , 16 , 21 , 22 ]'),\n",
            "((1247, 1250), 'all of which'),\n",
            "((1247, 1251), 'all of which used'),\n",
            "((1247, 1252), 'all of which used joints'),\n",
            "((1247, 1256), 'all of which used joints as the target tissue'),\n",
            "((1247, 1257), 'all of which used joints as the target tissue [ 15 , 16 , 21 , 22 ]'),\n",
            "((1250, 1251), 'used'),\n",
            "((1251, 1252), 'joints'),\n",
            "((1253, 1256), 'the target tissue'),\n",
            "((1256, 1257), '[ 15 , 16 , 21 , 22 ]'),\n",
            "((1256, 1257), '15'),\n",
            "((1256, 1257), '16'),\n",
            "((1256, 1257), '21'),\n",
            "((1256, 1257), '22'),\n",
            "((1256, 1257), ']'),\n",
            "((1257, 1259), 'This study'),\n",
            "((1257, 1263), 'This study , for the first time ,'),\n",
            "((1257, 1270), 'This study , for the first time , detected gene expression in LNs during CIA'),\n",
            "((1257, 1270), 'This study , for the first time , detected gene expression in LNs during CIA .'),\n",
            "((1257, 1270), 'This study , for the first time , detected gene expression in LNs during CIA . [SEP]'),\n",
            "((1263, 1264), 'detected'),\n",
            "((1263, 1270), 'detected gene expression in LNs during CIA'),\n",
            "((1264, 1268), 'gene expression in LNs'),\n",
            "((1264, 1270), 'gene expression in LNs during CIA'),\n",
            "((1267, 1268), 'LN'),\n",
            "((1267, 1268), 'LNs'),\n",
            "((1270, 1272), 'We present'),\n",
            "((1270, 1294), 'We present an extensive study of gene expression patterns in LNs of both genetically susceptible and resistant strains at four different phases of CIA'),\n",
            "((1271, 1272), 'present'),\n",
            "((1272, 1288), 'an extensive study of gene expression patterns in LNs of both genetically susceptible and resistant strains'),\n",
            "((1272, 1294), 'an extensive study of gene expression patterns in LNs of both genetically susceptible and resistant strains at four different phases of CIA'),\n",
            "((1276, 1288), 'gene expression patterns in LNs of both genetically susceptible and resistant strains'),\n",
            "((1280, 1288), 'LNs of both genetically susceptible and resistant strains'),\n",
            "((1283, 1284), 'genetically'),\n",
            "((1283, 1285), 'genetically susceptible'),\n",
            "((1284, 1285), 'susceptible'),\n",
            "((1286, 1287), 'resistant'),\n",
            "((1289, 1294), 'four different phases of CIA'),\n",
            "((1297, 1300), 'differentially regulated genes'),\n",
            "((1304, 1324), 'the PI and CA phases , and only a small number of genes were differentially expressed in two or three phases'),\n",
            "((1310, 1316), 'only a small number of genes'),\n",
            "((1317, 1319), 'differentially expressed'),\n",
            "((1318, 1319), 'expressed'),\n",
            "((1320, 1323), 'two or three'),\n",
            "((1320, 1324), 'two or three phases'),\n",
            "((1324, 1325), 'This'),\n",
            "((1327, 1331), 'biological responses in LNs'),\n",
            "((1330, 1331), 'LNs'),\n",
            "((1334, 1344), 'the PI and CA phases than in the OA phase'),\n",
            "((1334, 1350), 'the PI and CA phases than in the OA phase , and the responses at different phases'),\n",
            "((1342, 1343), 'OA'),\n",
            "((1345, 1350), 'the responses at different phases'),\n",
            "((1354, 1360), 'the susceptible to the resistant strain'),\n",
            "((1354, 1360), 'the susceptible to the resistant strain ,'),\n",
            "((1354, 1374), 'the susceptible to the resistant strain , the biggest difference was found in one cluster of genes ( cluster IV , Figure 2 )'),\n",
            "((1355, 1356), 'susceptible'),\n",
            "((1357, 1360), 'the resistant strain'),\n",
            "((1358, 1359), 'resistant'),\n",
            "((1360, 1363), 'the biggest difference'),\n",
            "((1360, 1374), 'the biggest difference was found in one cluster of genes ( cluster IV , Figure 2 )'),\n",
            "((1364, 1365), 'found'),\n",
            "((1366, 1370), 'one cluster of genes'),\n",
            "((1369, 1370), 'genes'),\n",
            "((1370, 1374), 'cluster IV , Figure 2'),\n",
            "((1370, 1374), 'cluster IV , Figure 2 )'),\n",
            "((1372, 1374), 'Figure 2'),\n",
            "((1376, 1377), 'had'),\n",
            "((1377, 1385), 'a higher expression in DBA / 1 than in FVB / N'),\n",
            "((1377, 1395), 'a higher expression in DBA / 1 than in FVB / N at the early phases of CIA ( NC , PI and OA )'),\n",
            "((1381, 1385), 'DBA / 1 than in FVB / N'),\n",
            "((1386, 1395), 'the early phases of CIA ( NC , PI and OA )'),\n",
            "((1390, 1395), 'CIA ( NC , PI and OA )'),\n",
            "((1391, 1395), 'NC , PI and OA'),\n",
            "((1391, 1395), 'NC , PI and OA )'),\n",
            "((1396, 1400), 'the opposite expression pattern'),\n",
            "((1401, 1404), 'the CA phase'),\n",
            "((1404, 1407), 'GO term classification'),\n",
            "((1404, 1418), 'GO term classification revealed that these genes were related to lymphocyte proliferation and activation'),\n",
            "((1409, 1418), 'these genes were related to lymphocyte proliferation and activation'),\n",
            "((1412, 1413), 'related'),\n",
            "((1414, 1416), 'lymphocyte proliferation'),\n",
            "((1414, 1418), 'lymphocyte proliferation and activation'),\n",
            "((1420, 1425), 'lymphocytes in the DBA / 1 strain'),\n",
            "((1420, 1428), 'lymphocytes in the DBA / 1 strain are more activated'),\n",
            "((1420, 1434), 'lymphocytes in the DBA / 1 strain are more activated than those in the FVB / N strain'),\n",
            "((1420, 1434), 'lymphocytes in the DBA / 1 strain are more activated than those in the FVB / N strain .'),\n",
            "((1425, 1426), 'are'),\n",
            "((1435, 1437), 'this difference'),\n",
            "((1435, 1454), 'this difference is not CIA specific because the expression difference between the two strains existed in mice without immunisation'),\n",
            "((1442, 1449), 'the expression difference between the two strains'),\n",
            "((1442, 1454), 'the expression difference between the two strains existed in mice without immunisation'),\n",
            "((1449, 1450), 'existed'),\n",
            "((1449, 1454), 'existed in mice without immunisation'),\n",
            "((1451, 1452), 'mice'),\n",
            "((1451, 1454), 'mice without immunisation'),\n",
            "((1453, 1454), 'immunisation'),\n",
            "((1454, 1455), 'Additionally'),\n",
            "((1455, 1462), 'some genes related to the immune response'),\n",
            "((1455, 1476), 'some genes related to the immune response were upregulated in the DBA / 1 strain but not in the FVB / N strain during CIA'),\n",
            "((1457, 1458), 'related'),\n",
            "((1459, 1462), 'the immune response'),\n",
            "((1463, 1464), 'upregulated'),\n",
            "((1465, 1474), 'the DBA / 1 strain but not in the FVB / N strain'),\n",
            "((1465, 1476), 'the DBA / 1 strain but not in the FVB / N strain during CIA'),\n",
            "((1472, 1473), 'FVB / N'),\n",
            "((1476, 1478), 'These differences'),\n",
            "((1481, 1488), 'a higher antibody response to collagen II'),\n",
            "((1481, 1498), 'a higher antibody response to collagen II occurred in the DBA / 1 strain compared to the FVB / N strain'),\n",
            "((1486, 1488), 'collagen II'),\n",
            "((1488, 1489), 'occurred'),\n",
            "((1490, 1498), 'the DBA / 1 strain compared to the FVB / N strain'),\n",
            "((1493, 1494), 'compared'),\n",
            "((1502, 1509), 'the difference of the susceptibility to CIA'),\n",
            "((1502, 1512), 'the difference of the susceptibility to CIA between both strains'),\n",
            "((1502, 1512), 'the difference of the susceptibility to CIA between both strains .'),\n",
            "((1505, 1509), 'the susceptibility to CIA'),\n",
            "((1514, 1516), 'were identified'),\n",
            "((1514, 1539), 'were identified as potential candidate genes for six of the eight small - effect QTL according to their gene expression patterns during CIA and their genomic locations'),\n",
            "((1515, 1516), 'identified'),\n",
            "((1517, 1539), 'potential candidate genes for six of the eight small - effect QTL according to their gene expression patterns during CIA and their genomic locations'),\n",
            "((1521, 1527), 'six of the eight small - effect QTL'),\n",
            "((1521, 1539), 'six of the eight small - effect QTL according to their gene expression patterns during CIA and their genomic locations'),\n",
            "((1529, 1535), 'their gene expression patterns during CIA'),\n",
            "((1529, 1539), 'their gene expression patterns during CIA and their genomic locations'),\n",
            "((1536, 1539), 'their genomic locations'),\n",
            "((1543, 1544), 'located'),\n",
            "((1545, 1561), 'QTL 4 and 7 , suggesting that QTG polymorphisms of the susceptibility genes inside these two QTL'),\n",
            "((1546, 1547), '4'),\n",
            "((1548, 1549), '7'),\n",
            "((1551, 1561), 'QTG polymorphisms of the susceptibility genes inside these two QTL'),\n",
            "((1554, 1561), 'the susceptibility genes inside these two QTL'),\n",
            "((1564, 1566), 'the phenotype'),\n",
            "((1567, 1568), 'regulating'),\n",
            "((1568, 1570), 'gene expression'),\n",
            "((1570, 1576), 'Two of the 21 candidate genes'),\n",
            "((1570, 1583), 'Two of the 21 candidate genes were reported to be involved in arthritis'),\n",
            "((1572, 1576), 'the 21 candidate genes'),\n",
            "((1577, 1578), 'reported'),\n",
            "((1583, 1599), 'Mapk14 , a candidate gene for locus 8 and also called p38 mitogen - activated protein kinase ( MAPK ) alpha ,'),\n",
            "((1588, 1590), 'locus 8'),\n",
            "((1588, 1599), 'locus 8 and also called p38 mitogen - activated protein kinase ( MAPK ) alpha'),\n",
            "((1588, 1599), 'locus 8 and also called p38 mitogen - activated protein kinase ( MAPK ) alpha ,'),\n",
            "((1593, 1599), 'p38 mitogen - activated protein kinase ( MAPK ) alpha'),\n",
            "((1599, 1600), 'regulate'),\n",
            "((1599, 1600), 'regulates'),\n",
            "((1603, 1604), 'arthritis'),\n",
            "((1603, 1605), 'arthritis - essential cytokines'),\n",
            "((1603, 1612), 'arthritis - essential cytokines , such as tumour necrosis factor and interleukin - 1'),\n",
            "((1603, 1613), 'arthritis - essential cytokines , such as tumour necrosis factor and interleukin - 1 [ 32 ]'),\n",
            "((1607, 1610), 'tumour necrosis factor'),\n",
            "((1607, 1612), 'tumour necrosis factor and interleukin - 1'),\n",
            "((1611, 1612), 'interleukin - 1'),\n",
            "((1616, 1624), 'p38 MAPK could attenuate CIA in rats [ 33 ]'),\n",
            "((1619, 1620), 'attenuate'),\n",
            "((1620, 1623), 'CIA in rats'),\n",
            "((1620, 1624), 'CIA in rats [ 33 ]'),\n",
            "((1622, 1623), 'rats'),\n",
            "((1625, 1636), 'p38 MAPK is becoming a potential therapeutic target in RA [ 32 ]'),\n",
            "((1634, 1636), 'RA [ 32 ]'),\n",
            "((1636, 1643), 'Stat5a , a candidate gene for loci 6'),\n",
            "((1636, 1653), 'Stat5a , a candidate gene for loci 6 , is an essential molecule for lymphoid development and differentiation [ 34 ]'),\n",
            "((1637, 1643), 'a candidate gene for loci 6'),\n",
            "((1641, 1643), 'loci 6'),\n",
            "((1642, 1643), '6'),\n",
            "((1643, 1644), 'is'),\n",
            "((1644, 1652), 'an essential molecule for lymphoid development and differentiation'),\n",
            "((1648, 1652), 'lymphoid development and differentiation'),\n",
            "((1653, 1654), 'Stat5a - deficient'),\n",
            "((1653, 1655), 'Stat5a - deficient mice'),\n",
            "((1653, 1667), 'Stat5a - deficient mice were reported to lose tolerance , resulting in the development of autoimmune diseases'),\n",
            "((1656, 1657), 'reported'),\n",
            "((1658, 1659), 'lose'),\n",
            "((1665, 1667), 'autoimmune diseases'),\n",
            "((1667, 1684), 'Stat5a is suggested to contribute to tolerance through maintenance of the CD4 + CD25 + regulatory T cell population [ 35 ]'),\n",
            "((1668, 1669), 'is'),\n",
            "((1669, 1670), 'suggested'),\n",
            "((1671, 1672), 'contribute'),\n",
            "((1673, 1674), 'tolerance'),\n",
            "((1675, 1684), 'maintenance of the CD4 + CD25 + regulatory T cell population [ 35 ]'),\n",
            "((1677, 1683), 'the CD4 + CD25 + regulatory T cell population'),\n",
            "((1677, 1684), 'the CD4 + CD25 + regulatory T cell population [ 35 ]'),\n",
            "((1684, 1685), 'Therefore'),\n",
            "((1685, 1687), 'we have'),\n",
            "((1685, 1688), 'we have presented'),\n",
            "((1685, 1702), 'we have presented a strategy to identify small - effect QTL and search for potential candidate genes within them .'),\n",
            "((1687, 1688), 'presented'),\n",
            "((1688, 1694), 'a strategy to identify small - effect QTL'),\n",
            "((1688, 1702), 'a strategy to identify small - effect QTL and search for potential candidate genes within them'),\n",
            "((1688, 1702), 'a strategy to identify small - effect QTL and search for potential candidate genes within them .'),\n",
            "((1691, 1692), 'identify'),\n",
            "((1691, 1694), 'identify small - effect QTL'),\n",
            "((1695, 1696), 'search'),\n",
            "((1703, 1718), 'it is noteworthy that the low statistical threshold and small number of animals per group'),\n",
            "((1703, 1720), 'it is noteworthy that the low statistical threshold and small number of animals per group could lead'),\n",
            "((1707, 1711), 'the low statistical threshold'),\n",
            "((1707, 1718), 'the low statistical threshold and small number of animals per group'),\n",
            "((1709, 1711), 'statistical threshold'),\n",
            "((1712, 1718), 'small number of animals per group'),\n",
            "((1715, 1716), 'animals'),\n",
            "((1717, 1718), 'group'),\n",
            "((1718, 1720), 'could lead'),\n",
            "((1719, 1720), 'lead'),\n",
            "((1726, 1729), 'the genome level'),\n",
            "((1729, 1745), 'some of the eight small - effect QTL identified using a very low threshold value ( P < 0 . 05 )'),\n",
            "((1729, 1747), 'some of the eight small - effect QTL identified using a very low threshold value ( P < 0 . 05 ) could be'),\n",
            "((1735, 1736), 'identified'),\n",
            "((1737, 1745), 'a very low threshold value ( P < 0 . 05 )'),\n",
            "((1745, 1746), 'could'),\n",
            "((1745, 1747), 'could be'),\n",
            "((1746, 1747), 'be'),\n",
            "((1749, 1770), 'For example , locus 4 was identified with a low P value and does not overlap with any previously identified arthritis QTL . [SEP]'),\n",
            "((1751, 1753), 'locus 4'),\n",
            "((1751, 1770), 'locus 4 was identified with a low P value and does not overlap with any previously identified arthritis QTL'),\n",
            "((1751, 1770), 'locus 4 was identified with a low P value and does not overlap with any previously identified arthritis QTL .'),\n",
            "((1751, 1770), 'locus 4 was identified with a low P value and does not overlap with any previously identified arthritis QTL . [SEP]'),\n",
            "((1754, 1755), 'identified'),\n",
            "((1756, 1760), 'a low P value'),\n",
            "((1756, 1770), 'a low P value and does not overlap with any previously identified arthritis QTL'),\n",
            "((1768, 1770), 'arthritis QTL'),\n",
            "((1773, 1774), ','),\n",
            "((1774, 1781), 'the small number of animals per group'),\n",
            "((1774, 1790), 'the small number of animals per group and the low threshold used to detect gene expression'),\n",
            "((1774, 1801), 'the small number of animals per group and the low threshold used to detect gene expression could also result in false positives in the differentially expressed genes'),\n",
            "((1780, 1781), 'group'),\n",
            "((1782, 1790), 'the low threshold used to detect gene expression'),\n",
            "((1792, 1793), 'result'),\n",
            "((1797, 1801), 'the differentially expressed genes'),\n",
            "((1801, 1802), 'Furthermore'),\n",
            "((1802, 1808), 'the differential expression of a gene'),\n",
            "((1802, 1823), 'the differential expression of a gene could result not only from allele difference between two strains , but also from other factors .'),\n",
            "((1806, 1808), 'a gene'),\n",
            "((1809, 1810), 'result'),\n",
            "((1813, 1818), 'allele difference between two strains'),\n",
            "((1813, 1823), 'allele difference between two strains , but also from other factors'),\n",
            "((1813, 1823), 'allele difference between two strains , but also from other factors .'),\n",
            "((1824, 1826), 'our findings'),\n",
            "((1830, 1832), 'future studies'),\n",
            "((1832, 1833), 'Conclusion'),\n",
            "((1833, 1835), 'We present'),\n",
            "((1833, 1844), 'We present a strategy to search candidate genes for small - effect QTL'),\n",
            "((1834, 1835), 'present'),\n",
            "((1838, 1839), 'search'),\n",
            "((1842, 1844), 'small - effect QTL'),\n",
            "((1845, 1859), 'this strategy , we identified 21 candidate genes for 8 small - effect QTL regulating CIA susceptibility'),\n",
            "((1847, 1849), 'we identified'),\n",
            "((1847, 1859), 'we identified 21 candidate genes for 8 small - effect QTL regulating CIA susceptibility'),\n",
            "((1848, 1849), 'identified'),\n",
            "((1849, 1859), '21 candidate genes for 8 small - effect QTL regulating CIA susceptibility'),\n",
            "((1853, 1859), '8 small - effect QTL regulating CIA susceptibility'),\n",
            "((1857, 1859), 'CIA susceptibility'),\n",
            "((1859, 1862), 'Our future studies'),\n",
            "((1859, 1869), 'Our future studies will be carried out using two approaches'),\n",
            "((1867, 1869), 'two approaches'),\n",
            "((1869, 1871), 'The first'),\n",
            "((1873, 1891), 'congenic animals for promising small - effect QTL that have relatively high P values and overlap with previously - identified arthritis QTL'),\n",
            "((1873, 1891), 'congenic animals for promising small - effect QTL that have relatively high P values and overlap with previously - identified arthritis QTL .'),\n",
            "((1876, 1879), 'promising small - effect QTL'),\n",
            "((1876, 1891), 'promising small - effect QTL that have relatively high P values and overlap with previously - identified arthritis QTL'),\n",
            "((1888, 1891), 'previously - identified arthritis QTL'),\n",
            "((1889, 1891), 'arthritis QTL'),\n",
            "((1891, 1894), 'The second approach'),\n",
            "((1891, 1904), 'The second approach is investigating candidate genes using both mouse and human studies'),\n",
            "((1895, 1896), 'investigating'),\n",
            "((1896, 1898), 'candidate genes'),\n",
            "((1899, 1904), 'both mouse and human studies'),\n",
            "((1900, 1901), 'mouse'),\n",
            "((1904, 1919), 'Candidate genes will be selected according to their function and polymorphism between the two strains'),\n",
            "((1908, 1909), 'selected'),\n",
            "((1911, 1919), 'their function and polymorphism between the two strains'),\n",
            "((1919, 1920), 'Thereafter'),\n",
            "((1920, 1923), 'we will generate'),\n",
            "((1920, 1934), 'we will generate knock - out mice to investigate the role of the genes in CIA'),\n",
            "((1922, 1923), 'generate'),\n",
            "((1927, 1934), 'the role of the genes in CIA'),\n",
            "((1930, 1934), 'the genes in CIA'),\n",
            "((1935, 1947), 'the loci whose counterparts on the human genome are linked to RA'),\n",
            "((1940, 1943), 'the human genome'),\n",
            "((1944, 1945), 'linked'),\n",
            "((1947, 1950), 'we will investigate'),\n",
            "((1947, 1960), 'we will investigate the candidate genes using case - control association studies in RA cohorts'),\n",
            "((1949, 1950), 'investigate'),\n",
            "((1950, 1953), 'the candidate genes'),\n",
            "((1950, 1960), 'the candidate genes using case - control association studies in RA cohorts'),\n",
            "((1954, 1960), 'case - control association studies in RA cohorts'),\n",
            "((1958, 1960), 'RA cohorts'),\n",
            "((1960, 1961), 'Abbreviations'),\n",
            "((1961, 1962), 'CA'),\n",
            "((1961, 1965), 'CA = chronic arthritis'),\n",
            "((1961, 1974), 'CA = chronic arthritis ; CIA = collagen - induced arthritis ; CV = coefficiant of variation'),\n",
            "((1963, 1965), 'chronic arthritis'),\n",
            "((1967, 1969), 'collagen - induced arthritis'),\n",
            "((1967, 1974), 'collagen - induced arthritis ; CV = coefficiant of variation'),\n",
            "((1969, 1974), 'CV = coefficiant of variation'),\n",
            "((1971, 1974), 'coefficiant of variation'),\n",
            "((1974, 1978), 'GO = Gene Ontology'),\n",
            "((1976, 1978), 'Gene Ontology'),\n",
            "((1980, 1982), 'lymph node'),\n",
            "((1982, 1983), 'LOD'),\n",
            "((1984, 1988), 'logarithm of the odds'),\n",
            "((1988, 1989), 'MAPK'),\n",
            "((1988, 1997), 'MAPK = mitogen - activated protein kinase ; NC = naive control'),\n",
            "((1988, 2002), 'MAPK = mitogen - activated protein kinase ; NC = naive control ; OA = onset of arthritis'),\n",
            "((1990, 1993), 'mitogen - activated protein kinase'),\n",
            "((1993, 1994), 'NC'),\n",
            "((1995, 1997), 'naive control'),\n",
            "((2004, 2005), 'post - immunisation'),\n",
            "((2007, 2010), 'quantitative trait gene'),\n",
            "((2007, 2021), 'quantitative trait gene ; QTL = quantitative trait loci ; RA = rheumatoid arthritis . Competing interests [SEP]'),\n",
            "((2010, 2011), 'QTL'),\n",
            "((2010, 2015), 'QTL = quantitative trait loci'),\n",
            "((2010, 2021), 'QTL = quantitative trait loci ; RA = rheumatoid arthritis . Competing interests [SEP]'),\n",
            "((2012, 2015), 'quantitative trait loci'),\n",
            "((2015, 2021), 'RA = rheumatoid arthritis . Competing interests [SEP]'),\n",
            "((2017, 2019), 'rheumatoid arthritis'),\n",
            "((2017, 2019), 'rheumatoid arthritis .'),\n",
            "((2017, 2021), 'rheumatoid arthritis . Competing interests [SEP]'),\n",
            "((2019, 2021), 'Competing interests'),\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "WGJ5JWDUddtQ",
        "outputId": "5b11cd69-2d2d-4f4a-9b13-ce45f06748cc"
      },
      "source": [
        "!zip -r /content/out.zip /content/coref/data/out "
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "text": [
            "  adding: content/coref/data/out/ (stored 0%)\n",
            "  adding: content/coref/data/out/17696610_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17244351_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17590087_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17590087_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17608565_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17425782_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17244351_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17194222_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17447844_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17425782_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17194222_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17425782_4.json (deflated 85%)\n",
            "  adding: content/coref/data/out/17696610_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17608565_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17590087_4.json (deflated 85%)\n",
            "  adding: content/coref/data/out/17194222_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17244351_4.json (deflated 85%)\n",
            "  adding: content/coref/data/out/17696610_4.json (deflated 85%)\n",
            "  adding: content/coref/data/out/17608565_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17590087_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17447844_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17696610_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17608565_4.json (deflated 85%)\n",
            "  adding: content/coref/data/out/17425782_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17590087_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17244351_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17447844_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17447844_4.json (deflated 85%)\n",
            "  adding: content/coref/data/out/17608565_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17425782_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17244351_3.json (deflated 84%)\n",
            "  adding: content/coref/data/out/17447844_1.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17194222_0.json (deflated 82%)\n",
            "  adding: content/coref/data/out/17696610_2.json (deflated 83%)\n",
            "  adding: content/coref/data/out/17194222_4.json (deflated 85%)\n"
          ],
          "name": "stdout"
        }
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 17
        },
        "id": "sX8IJG2Sd7aZ",
        "outputId": "c8a0dc14-55ef-4383-cf6d-d6a075287cac"
      },
      "source": [
        "files.download(\"/content/out.zip\")"
      ],
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "application/javascript": [
              "\n",
              "    async function download(id, filename, size) {\n",
              "      if (!google.colab.kernel.accessAllowed) {\n",
              "        return;\n",
              "      }\n",
              "      const div = document.createElement('div');\n",
              "      const label = document.createElement('label');\n",
              "      label.textContent = `Downloading \"${filename}\": `;\n",
              "      div.appendChild(label);\n",
              "      const progress = document.createElement('progress');\n",
              "      progress.max = size;\n",
              "      div.appendChild(progress);\n",
              "      document.body.appendChild(div);\n",
              "\n",
              "      const buffers = [];\n",
              "      let downloaded = 0;\n",
              "\n",
              "      const channel = await google.colab.kernel.comms.open(id);\n",
              "      // Send a message to notify the kernel that we're ready.\n",
              "      channel.send({})\n",
              "\n",
              "      for await (const message of channel.messages) {\n",
              "        // Send a message to notify the kernel that we're ready.\n",
              "        channel.send({})\n",
              "        if (message.buffers) {\n",
              "          for (const buffer of message.buffers) {\n",
              "            buffers.push(buffer);\n",
              "            downloaded += buffer.byteLength;\n",
              "            progress.value = downloaded;\n",
              "          }\n",
              "        }\n",
              "      }\n",
              "      const blob = new Blob(buffers, {type: 'application/binary'});\n",
              "      const a = document.createElement('a');\n",
              "      a.href = window.URL.createObjectURL(blob);\n",
              "      a.download = filename;\n",
              "      div.appendChild(a);\n",
              "      a.click();\n",
              "      div.remove();\n",
              "    }\n",
              "  "
            ],
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {
            "tags": []
          }
        },
        {
          "output_type": "display_data",
          "data": {
            "application/javascript": [
              "download(\"download_6c38b125-e10c-4a63-96c6-bdd5b519d4ad\", \"out.zip\", 637595)"
            ],
            "text/plain": [
              "<IPython.core.display.Javascript object>"
            ]
          },
          "metadata": {
            "tags": []
          }
        }
      ]
    }
  ]
}